---
title: "Prepare_data"
output: html_document
date: "2024-01-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE, warning=FALSE, echo=FALSE}
## Load required libraries
library(tidyverse)
library(ape)
library(distillR)
```

```{r}
#Read counts
read_counts <- read_tsv("resources/DMB0032_counts.tsv")%>%
  dplyr::rename(genome = 1)
genomes <- read_counts$genome
```

```{r}
#Genome metadata
genome_metadata <- read_tsv("resources/DMB0032_mag_info.tsv")%>%
  arrange(match(genome, genomes))

#%>%
#  mutate_at(vars(genome,domain,phylum,class,order,family,genus,species), ~ str_replace(., "[dpcofgs]__", "")) 
```

```{r}
#Genome table
genome_coverage <- read_tsv("resources/DMB0032_coverage.tsv")%>%
  dplyr::rename(genome = 1)%>%
  arrange(match(genome, genomes))

read_counts_row <- column_to_rownames(read_counts, "genome")
min_genome_coverage <- 0.3
read_counts_coverage <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_genome_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]]))

readlength=150 #change if sequencing read length is different
genome_counts <- read_counts_coverage %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$mag_size / readlength) ))

genome_counts <- genome_counts %>%
  column_to_rownames(., "genome") %>% 
  select(which(!colSums(., na.rm=TRUE) %in% 0))%>%
  rownames_to_column(., "genome")

genome_counts_rel <- genome_counts %>%
  mutate_at(vars(-genome), ~ . / sum(.))
```

```{r}
#Sample metadata
sample_metadata <- read_tsv("resources/DMB0032_metadata.tsv")%>%
	dplyr::rename(sample=EHI_plaintext) 

sample_metadata$region <- gsub("Ittoqqortoormii", "Ittoqqortoormiit", sample_metadata$region) # correct region names

metadata_unified <- read.csv("/Users/elsa/Desktop/THESIS/sled_dog_metagenomics/resources/metadata_unified.csv")
metadata_unified_clean <- metadata_unified %>%
  filter(sample != "")
selected_columns <- metadata_unified_clean %>%
  dplyr::select(sample, Q_code, PR_results_short, PR_toxocaris_leonina, PR_sarcocystis, PR_Tania_serialis, PR_cryptosporidium, PR_giardia, sero_toxoplasma_gondii_PR, sero_trichincella_spp_PR, Huldscore, Eukaryota_., Cryptosporidium_., Giardia_., Sarcocystis_., Toxoplasma_gondii_., Toxacaris_., Trichinella_., Uncinaria_., Echinococcus_., Taenia_., Cestoda_., Nematoda_., Virus_., PARVO_Parvoviridae_., PARVO_Bocaparvovirus_., PARVO_Canine_Minute_., PARVO_Canine_bufavirus_., PARVO_Canine_protoparvovirus_., PARVO_Canine_parvovirus_., ADENO_Adenoviridae_., ADENO_Canine_mastadenovirus_., PAP_Papillomaviridae_., )

sample_metadata <- dplyr::left_join(sample_metadata, selected_columns, by = "sample") # use for remaining analysis
```

```{r}
#Bacteria tree
tree <- read.tree("resources/DMB0032.tree")
tree <- keep.tip(tree, tip = genomes) # keep only MAG tips
```

```{r func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Genome functions
kegg_table <- read_tsv("resources/DMB0032_merged_kegg.tsv")
```

```{r func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Genome functions
genome_annotations <- read_delim("resources/DMB0032_annotations.csv")%>% column_to_rownames(., "...1")
```

```{r}
#Genome functions distillation
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19))
```

```{r}
save(read_counts, read_counts_coverage, genome_counts, tree, genome_metadata, kegg_table, sample_metadata, genome_gifts, genome_annotations, file = "resources/data/data.Rdata")
```

