---
title: "Elsa-Brenner-Masters-Thesis-Analysis"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# 1 Set up working Environment

## 1.1 Load required libraries and set working directory

```{r all_libraries, echo=FALSE, message=FALSE, warning=FALSE}
## Load required libraries
library(ggnewscale)
library(ggtree)
library(R.utils)
library(tidyverse)
library(ape)
library(devtools)
library(ggplot2)
library(hillR)
library(spaa)
library(vegan)
library(Rhdf5lib)
library(phyloseq)
library(phytools)
library(microbiome) 
library(matrixStats)
library(microbiomeutilities) 
library(lme4)
library(MuMIn)
library(nlme)
library(knitr) 
library(kableExtra)
library(pairwiseAdonis)
library(sjPlot)
library(distillR) 
library(RColorBrewer)
library(reshape2)
library(ggpubr)
library(ggdendro)
library(grid)
library(gplots)
library(dendextend)
library(stringr) 
library(Rtsne) 
library(glue) 
library(ggtree)
library(ggrepel)
library(ggpubr)
library(ggnewscale)
library(ggtreeExtra)
library(DESeq2)
library(ANCOMBC)
library(pheatmap)
library(hilldiv2) 
library(janitor)
library(ggh4x)
library(dplyr)
```
```{r loaddata, comment="", echo=TRUE, message=FALSE, warning=FALSE}
workingdir="/Users/elsa/Desktop/THESIS/sled_dog_metagenomics"
setwd(workingdir)
load("resources/data/data.Rdata")
```
# 2 Data pre-processing

```{r summary, echo=TRUE, message=FALSE, warning=FALSE}
read_counts_row <- column_to_rownames(read_counts, "genome")
nsamples <- ncol(read_counts_row) # define number of samples, by reading from the counts table
metagenomic_bases <- sum(sample_metadata$metagenomic_bases)
host_bases <- sum(sample_metadata$host_bases)
discarded_bases <- sum(round(((sample_metadata$metagenomic_bases+sample_metadata$host_bases)/(1-sample_metadata$bases_lost_fastp_percent))-(sample_metadata$metagenomic_bases+sample_metadata$host_bases))) #define amount of discarded bases from quality filtering
total_bases <- discarded_bases + host_bases + metagenomic_bases
singlem_bases <- sum(sample_metadata$metagenomic_bases * sample_metadata$singlem_fraction) 
nmags <- nrow(read_counts_row) # establish number of MAGs for analysis

sequencing_depth <- colSums(read_counts_row)
sequencing_depth_sum <- sum(sequencing_depth)
sequencing_depth_mean <- mean(sequencing_depth)
sequencing_depth_sd <- sd(sequencing_depth)
```

## 2.1 General statistics

**Number of samples in total**
```{r nsamples, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ncol(read_counts_row) # 58 samples
```
**Number of MAGs**
The number of metagenome-assembled genomes (MAG) or draft bacterial genomes reconstructed from the metagenomic data.

```{r nmags, comment="", echo=FALSE, message=FALSE, warning=FALSE}
cat(nmags) # 555 MAGs
```
**Amount of total data (GB):**
The amount of total DNA data sequenced in gigabases (GB, one billion nucleotide bases).

```{r totalGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
total_bases <- discarded_bases + host_bases + metagenomic_bases
totalgb <- round(total_bases / 1000000000,2) # convert to GB 
cat(totalgb) # 333.71 GB produced
```
**Amount of discarded data (GB):**
The amount of data discarded due to low quality or lack of informativeness during data preprocesing. Discarding 5-15% of the produced data is within the expected range, due to formation of adaptor dimers, inclusion of adaptors in sequencing reads due to short insert sizes, low sequencing quality, etc.

```{r discardedGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
discardgb <- sum(round(((sample_metadata$metagenomic_bases+sample_metadata$host_bases)/(1-sample_metadata$bases_lost_fastp_percent))-(sample_metadata$metagenomic_bases+sample_metadata$host_bases)))/1000000000
 # quality filtering
cat(discardgb) # 10.37 GB discarded
```
**Amount of discarded data (in % of the raw data):**

```{r %discarded, comment="", echo=FALSE, message=FALSE, warning=FALSE}
discarddata <- round(discarded_bases / total_bases * 100,2) 
cat(discarddata) # 3.11% removed after quality filtering
```
**Amount of host data (GB):**
The amount of data mapped against the host genome. The percentage refers to the amount of data mapped to the host genome respect to quality-filtered data. Note that this value can be very variable depending on the biological features of the sample (e.g., anal swabs contain more host DNA than faeces) and the employed reference genome (e.g., the chances for mapping to the genome are lower as the distance between) the study species and the employed reference genome differ).

```{r hostGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
hostGB <- round(host_bases / 1000000000,2)
cat(hostGB) # 6.49 GB from host
```
**Amount of host data (% of the quality-filtered data):**

```{r host%, comment="", echo=FALSE, message=FALSE, warning=FALSE}
hostdata <- round(host_bases / (total_bases-discarded_bases) * 100,2)
cat(hostdata) # 2.01% of quality filtered data is from host
```
**Estimated prokaryotic data:** 
The amount and proportion of data belonging to prokayotic genomes respect to the total metagenomic fraction, as estimated from singleM analysis. Note that this is an estimation that relies on the genome sizes of genomes available in reference databases. If a given taxon is not properly represented, genome size estimations can be less accurate.

```{r prokaGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
prokaGB <- round(singlem_bases / 1000000000,2)
cat(prokaGB) # 293.14 GB prokaryotic data
```
**Estimated prokaryotic data (% of the metagenomic data):** 

```{r proka%, comment="", echo=FALSE, message=FALSE, warning=FALSE}
prokadata <- round(singlem_bases / (metagenomic_bases) * 100,2)
cat(prokadata) # 92.51% of metagenomic data is prokaryotic
```
**Amount of metagenomic data (GB):**
The amount of data mapped against the host genome. The percentage refers to the amount of data mapped to the host genome respect to quality-filtered data. Note that this value can be very variable depending on the biological features of the sample (e.g., anal swabs contain more host DNA than faeces) and the employed reference genome (e.g., the chances for mapping to the genome are lower as the distance between) the study species and the employed reference genome differ).

```{r metaGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
metaGB <- round(metagenomic_bases / 1000000000,2)
cat(metaGB) # 316.85 GB metagenomic data
```
**Amount of metagenomic data (% of the quality-filtered data):**

```{r meta%, comment="", echo=FALSE, message=FALSE, warning=FALSE}
metaperce <- round(metagenomic_bases / (total_bases-discarded_bases) * 100,2)
cat(metaperce) # 97.99%  quality filtered data is metagenomic
```
**Total mapped sequencing depth (million reads):**
The amount of reads (and nucleotide bases) that were mapped to the entire MAG catalogue. Note that the amount of bases is only an approximation estimated by multiplying the exact number of mapped reads by 250 bp.

```{r totalreads, comment="", echo=FALSE, message=FALSE, warning=FALSE}
totalreads <- round(sequencing_depth_sum / 1000000,2)
cat(totalreads) # 1860.26 million reads
```
**Total mapped sequencing depth (GB):**

```{r mappedGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
mappedGB <- round(sequencing_depth_sum / 1000000000 * 143,2)
cat(mappedGB) # 266.02 GB total sequencing depth
```
**Average mapped sequencing depth (million reads):** 
This is the average number of reads (and nucleotide bases) mapped to each sample. Note that the amount of bases is only an approximation estimated by multiplying the exact number of mapped reads by 250 bp.
```{r meanreads, comment="", echo=FALSE, message=FALSE, warning=FALSE}
meanreads <- round(sequencing_depth_mean / 1000000,2)
cat(meanreads) # 32.07 million reads avg seq depth
```
**Average mapped sequencing depth (GB):** 
```{r meanGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
meanGB <- round(sequencing_depth_mean / 1000000000 * 143,2)
cat(meanGB) # 4.59 GB avg seq depth
```

## 2.2 General MAG statistics

**Number of MAGs without species-level annotation**
```{r nonspe, comment="", echo=FALSE, message=FALSE, warning=FALSE}
nonspecies <- genome_metadata %>%
    filter(species == "s__") %>%
    nrow()
cat(nonspecies) #244 MAGs lacking species level annotation
```

**Percentage of MAGs without species-level annotation**

```{r sp_percet, comment="", echo=FALSE, message=FALSE, warning=FALSE}
perct <- nonspecies*100/nmags
cat(perct) # 43.96% of MAGs with no species level annotation
```

**Number of phyla**

```{r phyla, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_metadata %>%
  dplyr::select(phylum) %>%
  unique() %>%
  pull() %>%
    length()  # 12 Phyla
```

**Number of genera**

```{r genera, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_metadata %>%
  dplyr::select(genus) %>%
  unique() %>%
  pull() %>%
    length()  # 135 genera
```
# 3 MAG catalogue

## 3.1 MAG tree

```{r list_phyla, echo=FALSE, warning=FALSE}
ehi_phylum_colors <- read.table("/Users/elsa/Desktop/THESIS/sled_dog_metagenomics/resources/ehi_phylum_colors copy.tsv",sep="\t",header=T,comment.char = "")

# Arrange colors alphabetically
colors_alphabetic <- ehi_phylum_colors %>%
  dplyr::right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, tree$tip.label)) %>%
  dplyr::select(phylum, colors) %>%
  mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()
```
```{r circular_tree_prep, echo=TRUE, warning=FALSE, comments="", message=FALSE, results="hide"}
heatmap <- ehi_phylum_colors %>%
  dplyr::right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
	arrange(match(genome, tree$tip.label)) %>%
  dplyr::select(genome,phylum) %>%
	mutate(phylum = factor(phylum, levels = unique(phylum))) %>%
	column_to_rownames(var = "genome")

circular_tree <- force.ultrametric(tree,method="extend") %>%
	ggtree(., layout = 'circular', size = 0.3)

# Phylum ring
circular_tree <- gheatmap(circular_tree, heatmap, offset=0.65, width=0.1, colnames=FALSE) +
		scale_fill_manual(values=colors_alphabetic) +
		geom_tiplab2(size=1, hjust=-0.1) +
		theme(plot.margin = margin(0, 0, 0, 0), panel.margin = margin(0, 0, 0, 0))

# Flush color scale to enable a new color scheme in the next ring
circular_tree <- circular_tree + new_scale_fill()

# Add completeness ring
circular_tree <- circular_tree +
        new_scale_fill() +
        scale_fill_gradient(low = "#d1f4ba", high = "#f4baba") +
        geom_fruit(
                data=genome_metadata,
                geom=geom_bar,
                mapping = aes(x=completeness, y=genome, fill=contamination),
                offset = 0.55,
                orientation="y",
              stat="identity")

# Add genome-size ring
circular_tree <-  circular_tree +
        new_scale_fill() +
        scale_fill_manual(values = "#cccccc") +
        geom_fruit(
             data=genome_metadata,
             geom=geom_bar,
             mapping = aes(x=mag_size, y=genome),
                 offset = 0.05,
                 orientation="y",
         stat="identity")

#Plot circular tree
circular_tree
```

  
## 3.2 MAG quality
\
Overview of the taxonomy and genome characteristics of the MAGs.\
\
**Completeness:** completeness of the MAG according to CheckM assessment.\
**Contamination:** contamination or redundancy of the MAG according to CheckM assessment.\
**Size:** size of the MAG in megabases (MB, one million nucleotide bases).

```{r biplot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv")%>%
  dplyr::right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
  arrange(match(genome, tree$tip.label)) %>%
  dplyr::select(phylum, colors) %>%
  unique() %>%
  arrange(phylum) %>%
  dplyr::select(colors) %>%
  pull()

#Generate quality biplot
genome_biplot <- genome_metadata %>%
  dplyr::select(c(genome,domain,phylum,completeness,contamination,mag_size)) %>%
  arrange(match(genome, rev(tree$tip.label))) %>% #sort MAGs according to phylogenetic tree
  ggplot(aes(x=completeness,y=contamination,size=mag_size,color=phylum)) +
              geom_point(alpha=0.7) +
                    ylim(c(10,0)) +
                    scale_color_manual(values=phylum_colors) +
                labs(y= "Contamination", x = "Completeness") +
                    theme_classic() +
                  theme(legend.position = "none")

#Generate contamination boxplot
genome_contamination <- genome_metadata %>%
            ggplot(aes(y=contamination)) +
                    ylim(c(10,0)) +
                    geom_boxplot(colour = "#999999", fill="#cccccc") +
                    theme_void() +
                    theme(legend.position = "none",
                        axis.title.x = element_blank(),
                        axis.title.y = element_blank(),
                        axis.text.y=element_blank(),
                        axis.ticks.y=element_blank(),
                        axis.text.x=element_blank(),
                        axis.ticks.x=element_blank(),
                        plot.margin = unit(c(0, 0, 0.40, 0),"inches")) #add bottom-margin (top, right, bottom, left)

#Generate completeness boxplot
genome_completeness <- genome_metadata %>%
        ggplot(aes(x=completeness)) +
                xlim(c(50,100)) +
                geom_boxplot(colour = "#999999", fill="#cccccc") +
                theme_void() +
                theme(legend.position = "none",
                    axis.title.x = element_blank(),
                    axis.title.y = element_blank(),
                    axis.text.y=element_blank(),
                    axis.ticks.y=element_blank(),
                    axis.text.x=element_blank(),
                    axis.ticks.x=element_blank(),
                    plot.margin = unit(c(0, 0, 0, 0.50),"inches")) #add left-margin (top, right, bottom, left)

#Render composite figure
library(gridExtra)
grid.arrange(grobs = list(genome_completeness,genome_biplot,genome_contamination),
        layout_matrix = rbind(c(1,1,1,1,1,1,1,1,1,1,1,4),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3)))
```

  
## 3.3 Sequencing depth assessment

```{r data_fraction, warning=FALSE, echo=FALSE, fig.height=5}
# Calculate sequence fractions
sequence_fractions <- read_counts_row %>%
  rownames_to_column("Genome") %>%
  pivot_longer(-Genome, names_to = "sample", values_to = "value") %>%
  group_by(sample) %>%
  summarise(mags = sum(value)) %>%
	dplyr::left_join(sample_metadata, by = join_by(sample == sample))  %>%
	dplyr::select(sample,mags,metagenomic_bases,host_bases,bases_lost_fastp_percent) %>%
	mutate(mags_bases = mags*146) %>%
	mutate(lowqual_bases = ((metagenomic_bases+host_bases)/(1-bases_lost_fastp_percent))-(metagenomic_bases+host_bases)) %>%
	mutate(unmapped_bases = metagenomic_bases - mags_bases) %>%
	dplyr::select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases)

mags_bases_mean <- sequence_fractions %>%
    mutate(mags_bases = mags_bases / 1000000000) %>%
    dplyr::select(mags_bases) %>%
    pull() %>%
    mean()

sequence_fractions_pivot <- sequence_fractions %>%
	pivot_longer(!sample, names_to = "fraction", values_to = "value") %>%
	mutate(value = value / 1000000000) %>%
	mutate(fraction = factor(fraction, levels = c("lowqual_bases","host_bases","unmapped_bases","mags_bases")))

sequence_fractions_meta <- merge(sequence_fractions_pivot, sample_metadata, by="sample")
```

```{r data_fraction_barplot, echo=FALSE, warning=FALSE}
ggplot(sequence_fractions_meta, aes(x = sample, y = value, fill=fraction)) +
  geom_bar(position="stack", stat = "identity") +
  scale_fill_manual(name=NULL,
                    breaks=c("lowqual_bases","host_bases","unmapped_bases","mags_bases"),
                    labels=c("Low quality","Host","Unmapped", "MAGs"),
                    values=c("#CCCCCC","#178a94","#ee8080","#d03161")) +
  geom_hline(yintercept = mags_bases_mean, linetype = "dashed", color = "black") +
  facet_grid(~region, scale="free", space="free") + # grouped the samples by region
  labs(x = "Samples", y = "Amount of data (GB)") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size = 6),
        strip.text.x = element_text(size = 7, face = "bold"),
        strip.background = element_rect(colour=NA, fill=NA),
        panel.background = element_rect(fill = NA, color = "black"),
        panel.border = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black"),
        legend.position="bottom",
        legend.title=element_blank())
```

  
## 3.4 Estimated vs. mapped prokaryotic fraction
```{r est_pro, comment="", echo=FALSE, message=FALSE, warning=FALSE}
singlem_table <- sequence_fractions %>%
    mutate(mags_proportion = round((mags_bases / (mags_bases + unmapped_bases))*100,2)) %>%
    dplyr::left_join(sample_metadata, by = join_by(sample == sample))  %>%
    mutate(singlem_proportion = round(singlem_fraction*100,2)) %>%
    dplyr::select(sample,mags_proportion,singlem_proportion) %>%
    mutate(mags_proportion = ifelse(singlem_proportion == 0, 0, mags_proportion)) %>% #convert zeros to NA
    mutate(singlem_proportion = ifelse(singlem_proportion == 0, NA, singlem_proportion)) %>% #convert zeros to NA
    mutate(singlem_proportion = ifelse(singlem_proportion < mags_proportion, NA, singlem_proportion)) %>% #if singlem is smaller, then NA, to simplify plot
    mutate(singlem_proportion = ifelse(singlem_proportion > 100, 100, singlem_proportion)) #simplify
```

```{r singlem, comment="", echo=FALSE, message=FALSE, warning=FALSE}
singlem_table %>%
    pivot_longer(!sample, names_to = "proportion", values_to = "value") %>%
    mutate(proportion = factor(proportion, levels = c("mags_proportion","singlem_proportion"))) %>%
    ggplot(., aes(x = value, y = sample, color=proportion)) +
            geom_line(aes(group = sample), color = "#f8a538") +
            geom_point() +
            scale_color_manual(values=c("#52e1e8","#876b53")) +
            theme_classic() +
            labs(y = "Samples", x = "Prokaryotic fraction (%)") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "right")
```

  
### 3.5 Additional sequencing needed

```{r seq_needs, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# Define the aimed GBs for host and mapped metagenomic data
mags_bases_aim=2
host_bases_aim=5

sequence_fractions_required <- sequence_fractions %>%
    mutate(mags_bases = round(mags_bases / 1000000000,2)) %>%
    mutate(unmapped_bases = round(unmapped_bases / 1000000000,2)) %>%
    mutate(host_bases = round(host_bases / 1000000000,2)) %>%
    mutate(lowqual_bases = round(lowqual_bases / 1000000000,2)) %>%
    mutate(total_bases = mags_bases+unmapped_bases+host_bases+lowqual_bases) %>%
    mutate(mags_bases_fraction = mags_bases/total_bases) %>%
    mutate(mags_bases_difference = mags_bases_aim - mags_bases) %>%
    mutate(meta_required = round(mags_bases_difference / mags_bases_fraction,2)) %>%
    mutate(meta_required = ifelse(meta_required < 0, 0, meta_required)) %>%
    mutate(host_bases_fraction = host_bases/total_bases) %>%
    mutate(host_bases_difference = host_bases_aim - host_bases) %>%
    mutate(host_required = round(host_bases_difference / host_bases_fraction,2)) %>%
    mutate(host_required = ifelse(host_required < 0, 0, host_required)) %>%
    dplyr::select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases,meta_required,host_required)
```

```{r seq_needs_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
sequence_fractions_required %>%
    dplyr::select(sample,meta_required,host_required) %>%
    mutate(meta_required = ifelse(meta_required > 20, 21, meta_required)) %>%
    mutate(host_required = ifelse(host_required > 20, 21, host_required)) %>%
    pivot_longer(!sample, names_to = "requirement", values_to = "value") %>%
    mutate(requirement = factor(requirement, levels = c("host_required","meta_required"))) %>%
    ggplot(., aes(x = value, y = sample, fill=requirement, group=requirement)) +
        geom_bar(position="stack", stat = "identity") +
            scale_fill_manual(values=c("#178a94","#d03161")) +
            facet_wrap(~requirement, scales="free_x") +
            labs(x = "Amount of data (GB)", y = "Samples") +
            geom_vline(xintercept = 20, linetype = "dashed", color = "black") +
        theme_classic() +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "bottom")
```

  
## 3.6 Relative abundance

```{r barplot1, comment="", echo=TRUE, message=FALSE, warning=FALSE}
genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  dplyr::left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  dplyr::left_join(., sample_metadata, by = join_by(sample == sample)) %>% #append sample metadata
  ggplot(., aes(x=sample,y=count, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(y = "Relative abundance") +
  facet_grid(.~region,  scales="free_x")+
#    facet_nested(.~Individual+Sample_type,  scales="free_x") +
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x=element_blank(),
          axis.title.x = element_blank(),
          axis.ticks.x=element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black")) +
   labs(fill="Phylum")
```

  
# 4 Alpha diversity calculations

Diversity estimations for each sample.\
\
**Richness:** Number of MAGs per sample (after applying coverage filter).\
**Neutral diversity:** Hill number of q=1 (Shannon diversity), a diversity metric that accounts for richness and eveness (relative abundances) of the MAGs.\
**Phylogenetic diversity:** Phylogenetic Hill number of q=1, a diversity metric that accounts for richness and eveness (relative abundances), as well as phylogenetic relationships among MAGs.\
**Functional diversity:** Functional Hill number of q=1, a diversity metric that accounts for richness and eveness (relative abundances), as well as functional dissimilarities among MAGs.\

```{r div_libraries, comment="", echo=FALSE, message=FALSE, warning=FALSE}
library(hilldiv2) 
library(janitor)
library(ggh4x)
library(gridExtra)
library(ggpubr)
library(tidyverse)
library(ggplot2)
library(vegan)
library(dplyr)
library(tidyr)
library(distillR)
```

## 4.1 Alpha diversity means by sample type
```{r alpha_div, comment="", echo=TRUE, message=FALSE, warning=FALSE}
#Calculate Hill numbers
richness <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            dplyr::select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=0) %>% 
            t() %>% 
            as.data.frame() %>%
            dplyr::rename(richness=1) %>%
            rownames_to_column(var="sample")

neutral <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            dplyr::select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1) %>% 
            t() %>% 
            as.data.frame() %>%
            dplyr::rename(neutral=1) %>%
            rownames_to_column(var="sample")

phylogenetic <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            dplyr::select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1,tree=tree) %>% 
            t() %>% 
            as.data.frame() %>%
            dplyr::rename(phylogenetic=1) %>%
            rownames_to_column(var="sample")

# Aggregate basal GIFT into elements
dist <- genome_gifts %>%
    to.elements(., GIFT_db) %>%
    traits2dist(., method="gower")

genome_counts_filt <- genome_counts[genome_counts$genome %in% rownames(genome_gifts),] 
rownames(genome_counts_filt) <- NULL
functional <- genome_counts_filt %>% 
            column_to_rownames(var="genome") %>% 
            dplyr::select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1,dist=dist) %>% 
            t() %>% 
            as.data.frame() %>%
            dplyr::rename(functional=1) %>%
            rownames_to_column(var="sample") %>%
            mutate(functional = if_else(is.nan(functional), 1, functional))

# Merge all metrics
alpha_div <- richness %>%
      dplyr::full_join(neutral,by=join_by(sample==sample)) %>%
      dplyr::full_join(phylogenetic,by=join_by(sample==sample)) %>%
      dplyr::full_join(functional,by=join_by(sample==sample)) %>%
      dplyr::left_join(., sample_metadata, by = join_by(sample == sample))


richness_mean <- alpha_div %>%
  group_by(region) %>%
  dplyr::summarise_at(.vars = names(.)[2], .funs = c("Richness mean" = "mean", "Richness sd" = "sd")) %>%
  dplyr::rename("Region"="region")

neutral_mean <- alpha_div %>%
  group_by(region) %>%
  dplyr::summarise_at(.vars = names(.)[3], .funs = c("Neutral mean" = "mean", "Neutral sd" = "sd"))

phylogenetic_mean <- alpha_div %>%
  group_by(region) %>%
  dplyr::summarise_at(.vars = names(.)[4], .funs = c("Phylogenetic mean" = "mean", "Phylogenetic sd" = "sd"))

functional_mean <- alpha_div %>%
  group_by(region) %>%
  dplyr::summarise_at(.vars = names(.)[5], .funs = c("Functional mean" = "mean", "Functional sd" = "sd"))

cbind(richness_mean, neutral_mean[, 2:3], phylogenetic_mean[, 2:3], 
      functional_mean[, 2:3]) %>% 
  as.data.frame()%>%
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  knitr::kable(.,digits = c(3,3))
```

## 4.2 Plot alpha diversities
```{r alpha_div_plot1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
alpha_div_pivot <- richness %>%
  dplyr::full_join(neutral,by=join_by(sample==sample)) %>%
  dplyr::full_join(phylogenetic,by=join_by(sample==sample)) %>%
  dplyr::full_join(functional,by=join_by(sample==sample)) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  dplyr::left_join(., sample_metadata, by = join_by(sample == sample))

alpha_div_pivot %>%
  ggplot(aes(x=value, y=sample)) +
  geom_bar(stat='identity', fill="#6c9ebc") +
  facet_nested(region ~ data,  scales="free")+
  coord_cartesian(xlim = c(1, NA)) +
  theme_classic() +
  theme(strip.background = element_blank(),
    panel.grid.minor.x = element_line( size=.1, color="grey" ),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1))

```

```{r alpha_div_rich, comment="", echo=TRUE, message=FALSE, warning=FALSE}
alpha_colors <- c("#e5bd5b", "#6b7398")
group_n <- alpha_div %>%
  dplyr::select(region) %>%
  pull() %>%
  unique() %>%
  length()

# all alpha diversity plots follow the same structure, thus only the script for neutral alpha diversity is shown. 

plot1 <- alpha_div %>%
  ggplot(aes(x = region, y = richness, group = region, color = region, fill = region)) +
  geom_jitter(width = 0.05, size = 1.5, show.legend = FALSE) +
  geom_violin(alpha = 0.2, width = 0.3, show.legend = FALSE) +
  scale_color_manual(values = alpha_colors[c(1:group_n)]) +
  scale_fill_manual(values = paste0(alpha_colors[c(1:group_n)], "50")) +
  stat_compare_means(show.legend = FALSE) +
  theme(axis.text.x = element_text(vjust = 0.5),
    axis.title = element_text(size = 14, face = "bold"),
    axis.text = element_text(face = "bold", size = 12),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0)),
    panel.background = element_blank(),
    # panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black")) +
  guides(colour = guide_legend(override.aes = list(size = 6))) +
  labs(x = "Region", y = "Richness")
```
```{r alpha_div_neutral, comment="", echo=FALSE, message=FALSE, warning=FALSE}
plot2 <- alpha_div%>%
  ggplot(aes(x = region, y = neutral, group = region, color = region, fill = region)) +
  geom_jitter(width = 0.05, size = 1.5, show.legend = FALSE) +
  geom_violin(alpha = 0.2, width = 0.3, show.legend = FALSE) +
  scale_color_manual(values = alpha_colors[c(1:group_n)]) +
  scale_fill_manual(values = paste0(alpha_colors[c(1:group_n)], "50")) +
  stat_compare_means(show.legend = FALSE) +
  theme(axis.text.x = element_text(vjust = 0.5),
    axis.title = element_text(size = 14, face = "bold"),
    axis.text = element_text(face = "bold", size = 12),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0)),
    panel.background = element_blank(),
    # panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black")) +
  guides(colour = guide_legend(override.aes = list(size = 6)))+
  labs(x = "Region", y = "Neutral")
```
```{r alpha_div_phylo, comment="", echo=FALSE, message=FALSE, warning=FALSE}
plot3 <- alpha_div%>%
  ggplot(aes(x = region, y = phylogenetic, group = region, color = region, fill = region)) +
  geom_jitter(width = 0.05, size = 1.5, show.legend = FALSE) +
  geom_violin(alpha = 0.2, width = 0.3, show.legend = FALSE) +
  scale_color_manual(values = alpha_colors[c(1:group_n)]) +
  scale_fill_manual(values = paste0(alpha_colors[c(1:group_n)], "50")) +
  stat_compare_means(show.legend = FALSE) +
  theme(axis.text.x = element_text(vjust = 0.5),
    axis.title = element_text(size = 14, face = "bold"),
    axis.text = element_text(face = "bold", size = 12),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0)),
    panel.background = element_blank(),
    # panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black")) +
  guides(colour = guide_legend(override.aes = list(size = 6)))+
  labs(x = "Region", y = "Phylogenetic")
```
```{r alpha_div_func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
plot4 <- alpha_div%>%
  ggplot(aes(x = region, y = functional, group = region, color = region, fill = region)) +
  geom_jitter(width = 0.05, size = 1.5, show.legend = FALSE) +
  geom_violin(alpha = 0.2, width = 0.3, show.legend = FALSE) +
  scale_color_manual(values = alpha_colors[c(1:group_n)]) +
  scale_fill_manual(values = paste0(alpha_colors[c(1:group_n)], "50")) +
  stat_compare_means(show.legend = FALSE) +
  theme(axis.text.x = element_text(vjust = 0.5),
    axis.title = element_text(size = 14, face = "bold"),
    axis.text = element_text(face = "bold", size = 12),
    axis.title.y = element_text(margin = margin(t = 0, r = 20, b = 0, l = 0)),
    axis.title.x = element_text(margin = margin(t = 20, r = 0, b = 0, l = 0)),
    panel.background = element_blank(),
    # panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black")) +
  guides(colour = guide_legend(override.aes = list(size = 6)))+
  labs(x = "Region", y = "Functional")
```
```{r div_plot_together, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
# Combined plot
grid.arrange(arrangeGrob(plot1,plot2,plot3,
                         plot4, ncol = 2)) 
```

# 5 Beta diversity
```{r beta_div, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide', eval=FALSE}
beta_q0 <- genome_counts%>%
  column_to_rownames(., "genome")%>%
  hillpair(., q = 0)

beta_q1n <- genome_counts%>%
  column_to_rownames(., "genome")%>%
  hillpair(., q = 1)

beta_q1p <- genome_counts%>%
  column_to_rownames(., "genome")%>%
  hillpair(., q = 1, tree = tree)

beta_div_func <- genome_counts_filt%>%
  column_to_rownames(., "genome")%>%
  hillpair(.,q=1,dist=dist)

save(beta_q0,beta_q1n, beta_q1p, beta_div_func, file = "resources/data/beta_div.Rdata")
```
```{r beta_div_open, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
load("resources/data/beta_div.Rdata")
```

```{r betan, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
beta_metric <- beta_q1n$S

beta_q1n_nmds <- beta_metric %>%
  vegan::metaMDS(., trymax = 500, k = 2, verbosity = FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  dplyr::left_join(sample_metadata, by = join_by(sample == sample))

group_n <- length(unique(beta_q1n_nmds$region))
beta_colors <- c("#e5bd5b", "#6b7398")
```

## 5.1 Neutral beta diversity
```{r beta_div_neutral, comment="", echo=TRUE, message=FALSE, warning=FALSE}

# all beta diversity plots follow the same format, thus only the script for neutral beta diversity is shown. 

q1n_nmds <- beta_q1n_nmds %>%
  group_by(region) %>%
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup()
ggplot(q1n_nmds, aes(x = NMDS1, y = NMDS2, color = region)) +
  scale_color_manual(values = beta_colors[c(1:group_n)]) +
  scale_shape_manual(values = 1:10) +
  geom_point(size = 4) +
  #   stat_ellipse(aes(color = beta_q1n_nmds$Groups))+
  geom_segment(aes(x = x_cen, y = y_cen, xend = NMDS1, yend = NMDS2), alpha = 0.9) +
  theme_classic() +
  theme(
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12),
    axis.title = element_text(size = 20, face = "bold"),
    axis.text = element_text(face = "bold", size = 18),
    panel.background = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size = 18),
    legend.position = "right", legend.box = "vertical"
  )
```

***Homogeneity of variance***

```{r permu1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ps.disper.neutral <- betadisper(beta_metric, sample_metadata$region) 
permutest(ps.disper.neutral, pairwise = TRUE) 
```

***Permanova***
```{r adonisbeta2, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# Adonis
metarow <- column_to_rownames(sample_metadata, "sample")

adonis2(beta_metric ~ region+sex, data = metarow[labels(beta_metric), ], permutations = 999) %>%
  as.matrix() %>%
  knitr::kable(., digits = c(0,3,3,3,3))
```

```{r betap, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
beta_p_metric <- beta_q1p$S

beta_q1p_nmds <- beta_p_metric %>%
  vegan::metaMDS(., trymax = 500, k = 2, verbosity = FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  dplyr::left_join(sample_metadata, by = join_by(sample == sample))

group_n <- length(unique(beta_q1p_nmds$region))
beta_colors <- c("#e5bd5b", "#6b7398", "#76b183", "#d57d2c", "#2a2d26", "#f9d4cc", "#3c634e", "#ea68c3")
```

## 5.2 Phylogenetic beta diversity
```{r  beta_div_phylo, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# pdf(paste(workingdir,"/betaB692_caecum.pdf",sep=""),width=14, height=9)
q1p_nmds <- beta_q1p_nmds %>%
  group_by(region) %>%
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup()
ggplot(q1p_nmds, aes(x = NMDS1, y = NMDS2, color = region)) +
  scale_color_manual(values = beta_colors[c(1:group_n)]) +
  scale_shape_manual(values = 1:10) +
  geom_point(size = 4)  +
  geom_segment(aes(x = x_cen, y = y_cen, xend = NMDS1, yend = NMDS2), alpha = 0.9) +
  theme_classic() +
  theme(
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12),
    axis.title = element_text(size = 20, face = "bold"),
    axis.text = element_text(face = "bold", size = 18),
    panel.background = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size = 18),
    legend.position = "right", legend.box = "vertical"
  )
#+ geom_text(aes(label = Sample), size=8)
# dev.off()
```

***Homogeneity of variance***

```{r permu2, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ps.disper.phylo <- betadisper(beta_p_metric, sample_metadata$region) 
permutest(ps.disper.phylo, pairwise = TRUE) 
```

***Permanova***
```{r adonisbeta3, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#### Adonis
metarow <- column_to_rownames(sample_metadata, "sample")

adonis2(beta_p_metric ~ region+sex, data = metarow[labels(beta_p_metric), ], permutations = 999) %>%
  as.matrix() %>%
  knitr::kable(., digits = c(0,3,3,3,3))
```

## 5.3 Functional beta diversity
```{r betaf, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
beta_f_metric <- beta_div_func$S

beta_f_nmds <- beta_f_metric %>%
  vegan::metaMDS(., trymax = 500, k = 2, verbosity = FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  dplyr::left_join(sample_metadata, by = join_by(sample == sample))

group_n <- length(unique(beta_f_nmds$region))
beta_colors <- c("#e5bd5b", "#6b7398")
```

```{r  beta_div_func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# pdf(paste(workingdir,"/x.pdf",sep=""),width=14, height=9)
f_nmds <- beta_f_nmds %>%
  group_by(region) %>%
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup()
ggplot(f_nmds, aes(x = NMDS1, y = NMDS2, color = region)) +
  scale_color_manual(values = beta_colors[c(1:group_n)]) +
  scale_shape_manual(values = 1:10) +
  geom_point(size = 4) +
  #   stat_ellipse(aes(color = beta_q1n_nmds$Groups))+
  geom_segment(aes(x = x_cen, y = y_cen, xend = NMDS1, yend = NMDS2), alpha = 0.9) +
  theme_classic() +
  theme(
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12),
    axis.title = element_text(size = 20, face = "bold"),
    axis.text = element_text(face = "bold", size = 18),
    panel.background = element_blank(),
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size = 18),
    legend.position = "right", legend.box = "vertical"
  )
#+ geom_text(aes(label = Sample), size=8)
# dev.off()
```

***Homogeneity of variance***
```{r permu3, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ps.disper.functional <- betadisper(beta_f_metric, sample_metadata$region) 
permutest(ps.disper.functional, pairwise = TRUE) 
```

***Permanova***
```{r adonisbeta4, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#### Adonis

metarow <- column_to_rownames(sample_metadata, "sample")

adonis2(beta_f_metric ~ region, data = metarow[labels(beta_f_metric), ], permutations = 999) %>%
  as.matrix() %>%
  knitr::kable(., digits = c(0,3,3,3,3))
```

# 6 Functional capacity analysis 
```{r filter_annotations, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
#### read table ####
filtered_df <- read.table("/Users/elsa/Desktop/THESIS/sled_dog_metagenomics/resources/DMB0032_annotations.csv", sep="")
```

## 6.1 Functional GIFT distillation:
```{r gifts, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide', cache=TRUE}
#Aggregate bundle-level GIFTs into the compound level
GIFTs_elements <- to.elements(genome_gifts,GIFT_db)
GIFTs_elements_filtered <- GIFTs_elements[rownames(GIFTs_elements) %in% genome_counts_filt$genome,]
GIFTs_elements_filtered <- as.data.frame(GIFTs_elements_filtered) %>% 
  select_if(~ !is.numeric(.) || sum(.) != 0)

#Aggregate element-level GIFTs into the function level
GIFTs_functions <- to.functions(GIFTs_elements_filtered,GIFT_db)

#Aggregate function-level GIFTs into overall Biosynthesis, Degradation and Structural GIFTs
GIFTs_domains <- to.domains(GIFTs_functions,GIFT_db)

#Get community-weighed average GIFTs per sample
genome_counts_row <- genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% 
  column_to_rownames(., "genome") 

GIFTs_elements_community <- to.community(GIFTs_elements_filtered,genome_counts_row,GIFT_db)
GIFTs_functions_community <- to.community(GIFTs_functions,genome_counts_row,GIFT_db)
GIFTs_domains_community <- to.community(GIFTs_domains,genome_counts_row,GIFT_db)
```

### 6.2 Community elements differences
```{r comunity_elem, comment="", echo=TRUE, message=FALSE, warning=FALSE}
element_gift <- GIFTs_elements_community %>% 
  as.data.frame() %>% 
  rownames_to_column(., "sample") %>% 
  merge(., sample_metadata[c(1,4)], by="sample")
```

```{r commun_wilcox_elem, comment="", echo=TRUE, message=FALSE, warning=FALSE}
uniqueGIFT_db<- unique(GIFT_db[c(2,4,5,6)]) %>% unite("Function",Function:Element, sep= "_", remove=FALSE)

significant_elements <- element_gift %>%
    pivot_longer(-c(sample,region), names_to = "trait", values_to = "value") %>%
    group_by(trait) %>%
    summarise(p_value = wilcox.test(value ~ region, exact = FALSE)$p.value) %>%
    mutate(p_adjust=p.adjust(p_value, method="BH")) %>%
    filter(p_adjust < 0.05)%>%
  rownames_to_column(., "Elements")  %>%
  dplyr::left_join(.,uniqueGIFT_db[c(1,3)],by = join_by(trait == Code_element))

element_gift_t <- element_gift  %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "trait")

element_gift_filt <- subset(element_gift_t, trait %in% significant_elements$trait) %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "sample")%>% 
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))


element_gift_filt_means <- element_gift_filt %>%
  dplyr::select(-sample)%>%
  group_by(region)  %>%
  summarise(across(everything(), mean))%>%
   t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "Elements")  %>%
  dplyr::left_join(.,uniqueGIFT_db[c(1,3)],by = join_by(Elements == Code_element))
```

```{r elements_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
element_gift_names <- element_gift_filt%>%
  dplyr::select(-region)%>%
   t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "Elements")  %>%
  dplyr::left_join(.,uniqueGIFT_db[c(1,3)],by = join_by(Elements == Code_element))%>%
  dplyr::select(-Elements)%>%
  dplyr::select(Function, everything())%>%
   t()%>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "sample")%>% 
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))


colNames <- names(element_gift_names)[2:4] # change to change number of plotted elements
for(i in colNames){
  plt <- ggplot(element_gift_names, aes(x=region, y=.data[[i]], color = region)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt)
}
```

  
only three plots are shown to illustrate the plotting function. 

```{r sig_elements, echo=FALSE, message=FALSE, warning=FALSE}
merged_element <- significant_elements %>%
  merge(., element_gift_filt_means, by = "Function")

# library("writexl")
# write_xlsx(merged_element, "/Users/elsa/Desktop/THESIS/sled_dog_metagenomics\\elements.xlsx")
```


## 6.3 Community functions differences

```{r comunity_func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
function_gift <- GIFTs_functions_community %>% 
  as.data.frame() %>% 
  rownames_to_column(., "sample") %>% 
  merge(., sample_metadata[c(1,4)], by="sample")
```

```{r commun_wilcox_func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
unique_funct_db<- GIFT_db[c(3,4,5)] %>% 
  distinct(Code_function, .keep_all = TRUE)

significant_functional <- function_gift %>%
    pivot_longer(-c(sample,region), names_to = "trait", values_to = "value") %>%
    group_by(trait) %>%
    summarise(p_value = wilcox.test(value ~ region, exact = FALSE)$p.value) %>%
    mutate(p_adjust=p.adjust(p_value, method="BH")) %>%
    filter(p_adjust < 0.05)%>%
  dplyr::left_join(.,unique_funct_db[c(1,3)],by = join_by(trait == Code_function))
```

```{r  function_sig, comment="", echo=FALSE, message=FALSE, warning=FALSE}
function_gift_t <- function_gift  %>% 
  dplyr::select(-region)  %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "trait")

function_gift_filt <- subset(function_gift_t, trait %in% significant_functional$trait) %>% 
  t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "sample")%>% 
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))


function_gift_filt %>%
  dplyr::select(-sample)%>%
  group_by(region)  %>%
  summarise(across(everything(), mean))%>%
   t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "Code_function")  %>%
  dplyr::left_join(.,unique_funct_db[c(1,3)],by = join_by(Code_function == Code_function)) %>%
  knitr::kable(., digits = c(0,3,3,3,3))
```
```{r function_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
function_gift_names <- function_gift_filt%>%
  dplyr::select(-region)%>%
   t() %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "Code_function")  %>%
  dplyr::left_join(.,unique_funct_db[c(1,3)],by = join_by(Code_function == Code_function))%>%
  dplyr::select(-Code_function)%>%
  dplyr::select(Function, everything())%>%
   t()%>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric)  %>%
  rownames_to_column(., "sample")%>% 
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))


colNames <- names(function_gift_names)[2:4] # change to change number of plotted functions
for(i in colNames){
  plt <- ggplot(function_gift_names, aes(x=region, y=.data[[i]], color = region)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt)
}
```

  
only 3 plots are shown to illustrate the plotting function. 

### 6.4 Community domains differences

***No differences***
```{r comunity_dom, comment="", echo=FALSE, message=FALSE, warning=FALSE}
domain_gift <- GIFTs_domains_community %>% 
  as.data.frame() %>% 
  rownames_to_column(., "sample") %>% 
  merge(., sample_metadata[c(1,4)], by="sample")
```

```{r commun_wilcox_dom, comment="", echo=FALSE, message=FALSE, warning=FALSE}
unique_domain_db<- GIFT_db[c(4)] %>% 
  distinct(Domain, .keep_all = TRUE)

significant_domain <- domain_gift %>%
    pivot_longer(-c(sample,region), names_to = "trait", values_to = "value") %>%
    group_by(trait) %>%
    summarise(p_value = wilcox.test(value ~ region, exact = FALSE)$p.value) %>%
    mutate(p_adjust=p.adjust(p_value, method="BH"))
```

## 6.5 Functional capacity of the MAGs
\
***element***
```{r GIFTs_elements, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.dim = c(20, 20)}
GIFTs_elements %>%
  reshape2::melt() %>%
  dplyr::rename(Genome = Var1, Code_element = Var2, GIFT = value) %>%
  dplyr::inner_join(GIFT_db,by="Code_element") %>%
  ggplot(., aes(x=Code_element, y=Genome, fill=GIFT, group=Code_function))+
  geom_tile()+
  scale_y_discrete(guide = guide_axis(check.overlap = TRUE))+
  scale_x_discrete(guide = guide_axis(check.overlap = TRUE))+
  scale_fill_gradientn(limits = c(0,1), colours=brewer.pal(7, "YlGnBu"))+
  facet_grid(. ~ Code_function, scales = "free", space = "free")+
  theme_grey(base_size=8)+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),strip.text.x = element_text(angle = 90))
```
\
***function***
```{r hierachical_clustering, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.dim = c(20, 20)}
col <- colorRampPalette(brewer.pal(7, "YlGnBu"))(256)
heatmap(GIFTs_functions_community, scale = "none", col=col, Colv=NA,
        trace = "none", density.info = "none", ylab="Samples", 
        xlab="Gifts", cexRow=0.7, margins = c(8,8), lhei=c(2,4), lwid=c(2,5), 
        keysize=0.75, key.par = list(cex=0.9),srtCol = 45)
```

# 7 Differential abundance analysis
## 7.1 Remove structural zeros
```{r remove structural zeros, echo=TRUE, message=FALSE, warning=FALSE}
# remove MAGs only present in one region
Daneborg_samples <- sample_metadata %>% 
                    filter(region == "Daneborg") %>% 
                    dplyr::select(sample) %>% pull()

Ittoqqortoormiit_samples <- sample_metadata %>% 
                    filter(region == "Ittoqqortoormiit") %>% 
                    dplyr::select(sample) %>% pull()

structural_zeros <- genome_counts %>% 
   dplyr::rowwise() %>% #compute for each row (genome)
   mutate(all_zeros_Ittoqqortoormiit = all(c_across(all_of(Ittoqqortoormiit_samples)) == 0)) %>% # set true if all samples in Ittoqqortoormiit have zeros
   mutate(all_zeros_Daneborg = all(c_across(all_of(Daneborg_samples)) == 0)) %>% # set true if all samples in Daneborg have zeros
   mutate(average_Ittoqqortoormiit = mean(c_across(all_of(Ittoqqortoormiit_samples)), na.rm = TRUE)) %>% # get average genome counts across Ittoqqortoormiit
   mutate(average_Daneborg = mean(c_across(all_of(Daneborg_samples)), na.rm = TRUE)) %>% # get average genome counts across Daneborg
   filter(all_zeros_Ittoqqortoormiit == TRUE || all_zeros_Daneborg==TRUE)  %>% # filter only genomes with structural zeros
   mutate(present = case_when(
      all_zeros_Ittoqqortoormiit & !all_zeros_Daneborg ~ "Daneborg",
      !all_zeros_Ittoqqortoormiit & all_zeros_Daneborg ~ "Ittoqqortoormiit",
      !all_zeros_Ittoqqortoormiit & !all_zeros_Daneborg ~ "None",
      TRUE ~ NA_character_
    )) %>%
   mutate(average = ifelse(present == "Ittoqqortoormiit", average_Ittoqqortoormiit, average_Daneborg)) %>%
   dplyr::select(genome, present, average) %>%
   dplyr::left_join(genome_metadata, by=join_by(genome==genome)) %>%
   arrange(present,-average)
```
```{r plot structural zeros, echo=FALSE, message=FALSE, warning=FALSE}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(structural_zeros, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

structural_zeros %>%
    mutate(average = ifelse(present == "Ittoqqortoormiit", average * -1, average)) %>% #convert Ittoqqortoormiit genome counts to negative
    ggplot(., aes(x=average, y=forcats::fct_rev(phylum), color=phylum)) +
      geom_jitter(size=3) +
      geom_vline(xintercept=0) + 
      xlim(-max(structural_zeros$average)-3,max(structural_zeros$average)+3) +
      scale_color_manual(values=phylum_colors) +
      geom_text(aes(-max(structural_zeros$average)+5, 1), label = "IT", color="#666666") +
      geom_text(aes(max(structural_zeros$average)-5, 1), label = "DB", color="#666666") +
     theme(legend.position='none',
          panel.background = element_blank(),
          axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"))+
      labs(y="Genus",x="Genome counts") + 
      guides(col=guide_legend("Phylum"))
```

\
***Create phyloseq object (without structural zeros)***
```{r physeq object with normalized counts, structural zeros removed}
#phyloseq object without structural zeros

# OTU table
phylo_counts <- genome_counts %>% # genome size normalized counts
  dplyr::filter(!genome %in% structural_zeros$genome) %>% # remove structural zeros
  column_to_rownames("genome") %>%
  mutate_all(~ replace(., . == 0, 0.00001)) %>%
  otu_table(., taxa_are_rows = TRUE)

phylo_samples <- sample_metadata %>% 
                    column_to_rownames("sample") %>% 
                    sample_data() #convert to phyloseq sample_data object

# Tax table
phylo_taxonomy <- genome_metadata %>% 
                    filter(genome %in% rownames(phylo_counts)) %>% # remove structural zeros
                    mutate(genome2=genome) %>% #create a pseudo genome name column
                    column_to_rownames("genome2") %>% 
                    dplyr::select(domain,phylum,class,order,family,genus,species,genome) %>% #add an additional taxonomic level to ensure genome-level analysis (as no all genomes have species-level taxonomic assignments. Otherwise, ANCOMBC2 aggregates analyses per species)
                    as.matrix() %>% 
                    tax_table() #convert to phyloseq tax_table object 

# Tree
tree <- phytools::force.ultrametric(tree, method = "extend")
phylo_tree = phyloseq::phy_tree(tree)

#Generate phyloseq object required to input ANCOMBC
genome_data <- phyloseq(phylo_counts, phylo_taxonomy, phylo_samples, phylo_tree)
```

```{r clr_phylo, comment="", echo=TRUE, message=FALSE, warning=FALSE}
#clr transformation
physeq_clr <- microbiome::transform(genome_data, 'clr')

count_crl <- data.frame(physeq_clr@otu_table)
count_crl_t <- data.frame(t(count_crl))
metadata_crl <- data.frame(physeq_clr@sam_data)
metadata_crl <- rownames_to_column(metadata_crl, "sample")
metadata_crl <- metadata_crl[match(rownames(count_crl_t),metadata_crl$sample),]
design <- metadata_crl[, c("sample", "region")]
```

### 7.1.2 MAGs in different between location and shared between locations
```{r upset, comment="", echo=TRUE, message=FALSE, warning=FALSE}
## Upset visualization: MAGs in different locations and shared among locations
library(UpSetR)
locationcolors=c('#c4d7d1','#408892')

genome_counts_row<-genome_counts %>%column_to_rownames(., "genome")
Ittoqqortoormiit <- genome_counts_row[, colnames(genome_counts_row) %in% Ittoqqortoormiit_samples]%>%
  rowSums()%>% as.data.frame()%>% dplyr::rename(Ittoqqortoormiit=".")%>%rownames_to_column(., "genome")
  
table_upset_analysis <- genome_counts_row[, colnames(genome_counts_row) %in% Daneborg_samples]%>%
  rowSums()%>% as.data.frame()%>% dplyr::rename(Daneborg=".")%>%rownames_to_column(., "genome")%>%
      dplyr::left_join(., Ittoqqortoormiit, by = join_by(genome == genome))%>%column_to_rownames(., "genome")

table_upset_analysis_binary <- ifelse(table_upset_analysis > 0, 1, 0) %>% as.data.frame()

#pdf("figures/MAG_intersection.pdf",width=8,height=6, onefile=F)
upset(as.data.frame(table_upset_analysis_binary),
  keep.order = T,
#  sets = rev(c("feces","cloaca")),
  sets.bar.color= rev(locationcolors),
  mb.ratio = c(0.55, 0.45), order.by = "freq")
#dev.off()
```
***Total number only present in Ittoqqortoormiit and absent in Daneborg***
```{r speno_total_ittoq, comment="", echo=FALSE, message=FALSE, warning=FALSE}
no_daneborg <- table_upset_analysis_binary %>% dplyr::select(Daneborg) %>% dplyr::filter(.,Daneborg==0)%>%rownames_to_column(., "genome")%>%
      dplyr::left_join(., genome_metadata, by = join_by(genome == genome)) %>% dplyr::select(c(1,4:9))
total_absent <- no_daneborg %>%
	nrow()
total_absent # 88 Ittoqqortoormiit unique MAGs
```

***Mags only present in Ittoqqortoormiit with species-level annotation***
```{r spe_ittoq, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
no_daneborg%>%
	filter(species != "s__") # 45 Ittoqqortoormiit unique MAGs WITH species-level annotation
```

***Number of MAGs present only in Ittoqqortoormiit without species-level annotation***
```{r speno_perc_ittoq, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
new_species_absent <- no_daneborg %>%
	filter(species == "s__") %>%
	nrow()
no_daneborg %>%
	filter(species == "s__")%>% dplyr::select(c(1:6))
new_species_absent*100/total_absent # 43 Ittoqqortoormiit unique MAGs w/o species-level annotation
```

***Total number only present in Daneborg and absent in Ittoqqortoormiit***
```{r speno_total_daneb, comment="", echo=FALSE, message=FALSE, warning=FALSE}
no_Ittoqqortoormiit <- table_upset_analysis_binary %>% dplyr::select(Ittoqqortoormiit) %>% dplyr::filter(.,Ittoqqortoormiit==0)%>%rownames_to_column(., "genome")%>%
      dplyr::left_join(., genome_metadata, by = join_by(genome == genome)) %>% dplyr::select(c(1,4:9))
total_absent_Ittoqqortoormiit <- no_Ittoqqortoormiit %>%
	nrow()
cat(total_absent_Ittoqqortoormiit) # 37 MAGs only in Daneborg
```

***Mags only present in Daneborg with species-level annotation***

```{r spe_daneb, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
no_Ittoqqortoormiit%>%
	filter(species != "s__") # 16 Daneborg unique MAGs WITH species-level annotation
```

***Number of MAGs present only in Daneborg without species-level annotation***
```{r speno_perc_daneb, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
new_species_Ittoqqortoormiit <- no_Ittoqqortoormiit %>%
	filter(species == "s__") %>%
	nrow()
no_Ittoqqortoormiit %>%
	filter(species == "s__") %>% dplyr::select(c(1:6))
cat(new_species_Ittoqqortoormiit*100/total_absent_Ittoqqortoormiit) # 21 Daneborg unique MAGs w/o species-level annotation
```
## 7.2 Wilcoxon testing / (Mann Whitney U) 

***Daneborg phylum summary***
```{r phylum1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
physeq_phylum <- microbiome::aggregate_taxa(genome_data, 'phylum')
physeq_phylum_rel <-  microbiome::transform(physeq_phylum, "compositional")
physeq_phylum_rel <- subset_samples(physeq_phylum_rel, region == "Daneborg")
table.rel1 <- physeq_phylum_rel@otu_table*100
means.table.rel1 <- as.data.frame(rowMeans(table.rel1))
sd.table.rel1 <- as.data.frame(matrixStats::rowSds(table.rel1, useNames = TRUE))
summary.phylum1 <- merge(means.table.rel1, sd.table.rel1, by="row.names")
colnames(summary.phylum1) <- c("Phylum","Mean", "SD")
print(summary.phylum1[order(-summary.phylum1$Mean),], row.names = FALSE)
```

***Ittoqqortoormiit phylum summary***
```{r phylum2, comment="", echo=FALSE, message=FALSE, warning=FALSE}
physeq_phylum <- microbiome::aggregate_taxa(genome_data, 'phylum')
physeq_phylum_rel <-  microbiome::transform(physeq_phylum, "compositional")
physeq_phylum_rel_itto <- subset_samples(physeq_phylum_rel, region == "Ittoqqortoormiit")
table.rel1 <- physeq_phylum_rel_itto@otu_table*100
means.table.rel1 <- as.data.frame(rowMeans(table.rel1))
sd.table.rel1 <- as.data.frame(matrixStats::rowSds(table.rel1, useNames = TRUE))
summary.phylum1 <- merge(means.table.rel1, sd.table.rel1, by="row.names")
colnames(summary.phylum1) <- c("Phylum","Mean", "SD")
print(summary.phylum1[order(-summary.phylum1$Mean),], row.names = FALSE)
```

***Genome level***
\
both regional analyses are done in the same manner, thus only Ittoqqortoormiit is shown. 
\
**Ittoqqortoormiit**
```{r Ittoq summary for DA P, comment="", echo=TRUE, message=FALSE, warning=FALSE}
physeq_ittoq <- subset_samples(physeq_clr, region == "Ittoqqortoormiit")
physeq_ittoq <- prune_taxa(taxa_sums(physeq_ittoq)>0, physeq_ittoq)
table.rel1_ittoq <- physeq_ittoq@otu_table
means.table.rel1_ittoq <- as.data.frame(rowMeans(table.rel1_ittoq))
sd.table.rel1_ittoq <- as.data.frame(matrixStats::rowSds(table.rel1_ittoq, useNames = TRUE))
summary.ittoq <- merge(means.table.rel1_ittoq, sd.table.rel1_ittoq, by="row.names")
colnames(summary.ittoq) <- c("Genome","Mean", "SD")
head(summary.ittoq[order(-summary.ittoq$Mean),], row.names = FALSE)
```

**Daneborg**
```{r Daneborg summary for DA P, comment="", echo=FALSE, message=FALSE, warning=FALSE}
physeq_daneborg <- subset_samples(physeq_clr, region == "Daneborg")
physeq_daneborg <- prune_taxa(taxa_sums(physeq_daneborg)>0, physeq_daneborg)
table.rel1_daneborg <- physeq_daneborg@otu_table
means.table.rel1_daneborg <- as.data.frame(rowMeans(table.rel1_daneborg))
sd.table.rel1_daneborg <- as.data.frame(matrixStats::rowSds(table.rel1_daneborg, useNames = TRUE))
summary.daneborg <- merge(means.table.rel1_daneborg, sd.table.rel1_daneborg, by="row.names")
colnames(summary.daneborg) <- c("Genome","Mean", "SD")
head(summary.daneborg[order(-summary.daneborg$Mean),], row.names = FALSE)
```

***Genome***
```{r physeq_mag, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
physeq_genome_filtered_clr <- physeq_clr
```
```{r wilcox0_mag, comment="", echo=TRUE, message=FALSE, warning=FALSE}
clr_t <- as.data.frame(t(as.matrix(physeq_clr@otu_table))) %>%
  rownames_to_column(., "sample")%>% 
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))
significant_genome <- clr_t %>%
    pivot_longer(-c(sample,region), names_to = "Genome", values_to = "value") %>%
    group_by(Genome) %>%
    summarise(p_value = wilcox.test(value ~ region, exact = FALSE)$p.value) %>%
    mutate(p_adjust=p.adjust(p_value, method="BH")) %>%
    filter(p_adjust < 0.05)
significant_genome
```
```{r wilcox0_mag_plot, comment="", echo=TRUE, message=FALSE, warning=FALSE}
physeq_genome_filtered_rel <-  microbiome::transform(genome_data, "compositional")
genome <- as.data.frame(t(as.matrix(physeq_genome_filtered_rel@otu_table)))
significant_genome_table <- genome[, colnames(genome) %in% significant_genome$Genome] %>%
  rownames_to_column(., "sample")%>%  
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))%>%
  dplyr::select(-sample)

colNames <- names(significant_genome_table)[1:3] # change to change number of plotted MAGs
for(i in colNames){
  plt <- ggplot(significant_genome_table, aes(x=region, y=.data[[i]], color = region)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt)
}
```

  
only three plots are shown to illustrate the plotting function

***Genera level***
```{r physeq_genus, comment="", echo=FALSE, message=FALSE, warning=FALSE}
physeq_genome_filtered_genus <- microbiome::aggregate_taxa((physeq_clr), 'genus')
```
```{r wilcox0_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
physeq_clr_genus <- microbiome::aggregate_taxa(physeq_genome_filtered_clr, 'genus')
clr_t_genus <- as.data.frame(t(as.matrix(physeq_clr_genus@otu_table))) %>%
  rownames_to_column(., "sample")%>% 
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))
significant_genus <- clr_t_genus %>%
    pivot_longer(-c(sample,region), names_to = "Genus", values_to = "value") %>%
    group_by(Genus) %>%
    summarise(p_value = wilcox.test(value ~ region, exact = FALSE)$p.value) %>%
    mutate(p_adjust=p.adjust(p_value, method="BH")) %>%
    filter(p_adjust < 0.05)
significant_genus
```

```{r wilcox0_gen_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
physeq_genome_filtered_genus <- microbiome::aggregate_taxa((genome_data), 'genus')
physeq_genome_filtered_genus_rel <-  microbiome::transform(physeq_genome_filtered_genus, "compositional")
genus <- as.data.frame(t(as.matrix(physeq_genome_filtered_genus_rel@otu_table)))
significant_genus_table <- genus[, colnames(genus) %in% significant_genus$Genus] %>%
  rownames_to_column(., "sample")%>%  
  dplyr::left_join(., sample_metadata[c(1,4)], by = join_by(sample == sample))%>%
  dplyr::select(-sample)

colNames <- names(significant_genus_table)[1:3] # change to change number of plotted MAGs
for(i in colNames){
  plt <- ggplot(significant_genus_table, aes(x=region, y=.data[[i]], color = region)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt)
}
```

  
only three plots shown to illustrate the plotting function.

## 7.3 RDA
```{r prep_data, comment="", echo=FALSE, message=FALSE, warning=FALSE}
count_table_t <- data.frame(t(genome_counts)) %>%
  row_to_names(row_number = 1) %>%
  as.data.frame() %>%
  mutate_if(is.character, as.numeric) 
hel1 <- decostand(count_table_t, "hellinger")
```

```{r rda_model, comment="", echo=FALSE, message=FALSE, warning=FALSE}
rda.model <- rda(count_crl_t ~ region, data = design)# Run the model
rda.model.hel <- rda(hel1 ~ region, data = design)# Run the model
#summary(rda.model)
# Anova
anova(rda.model)
anova(rda.model.hel)
# anova(hel_rda,by="axis")
# summary(hel_rda)
RsquareAdj(rda.model)
RsquareAdj(rda.model.hel)
anova.cca(rda.model, step = 1000)
anova.cca(rda.model, step = 1000, by = "term")
anova.cca(rda.model, step = 1000, by = "axis")
perc <- round(100*(summary(rda.model)$cont$importance[2, 1:2]), 2)
## extract scores - these are coordinates in the RDA space
sc_si <- vegan::scores(rda.model, display="sites", choices=c(1,2), scaling=1)
sc_sp <- vegan::scores(rda.model, display="species", choices=c(1,2), scaling=1)
sc_bp <- vegan::scores(rda.model, display="bp", choices=c(1, 2), scaling=1)

# Prepare data for plotting
rda_sc_wa <- vegan::scores(rda.model,
  display = "wa",
  scaling = 1
)
rda_sc_cn <- data.frame(vegan::scores(rda.model,
  display = "cn",
  scaling = 1
))
rda_sc_sp <- data.frame(vegan::scores(rda.model,
  display = "sp",
  scaling = 2
))

# Select MAGs from quantile
ASV_Scores_RDA1quantiles <- quantile(rda_sc_sp$RDA1,
  probs = c(0.01, 0.99)
)
HighFit_ASVs <- rda_sc_sp[
  rda_sc_sp$RDA1 < ASV_Scores_RDA1quantiles[1] |
    rda_sc_sp$RDA1 > ASV_Scores_RDA1quantiles[2],
]

HighFit_ASVNames <- rownames(HighFit_ASVs)
species_index <- data.frame(
  Species = HighFit_ASVNames,
  Index = 1:length(HighFit_ASVNames)
)

# combine dataframes
all_dataframe <- data.frame(rda_sc_wa, design)
```
```{r rda_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# Plot
ggplot(all_dataframe, aes(y = PC1, x = RDA1)) +
  geom_point(aes(y = PC1, x = RDA1, colour = region)) +
  # text
  xlab(paste0("RDA1 (", perc[1], "%)")) +
  ylab(paste0("PC1 (", perc[2], "%)")) +
  # segment
geom_segment(data = HighFit_ASVs,
             aes(x = 0, xend = RDA1, y = 0, yend = PC1),
             arrow = arrow(length = unit(0.25, "cm")),
             color = "grey") +
geom_point(data = HighFit_ASVs,
            aes(y = PC1, x = RDA1),
            show.legend = FALSE,
            color = "grey") +
geom_label_repel(data = HighFit_ASVs,
                  aes(y = PC1, x = RDA1,label = species_index$Species),
                  size = 3,
                  show.legend = FALSE,
                  colour = "grey",
                  max.overlaps = 20) +
#  scale_colour_manual(values = c("red", "blue")) +
  scale_shape_discrete(solid = TRUE) +
  theme_bw() +
  theme(
    plot.title = element_text(size = 10),
    legend.position = "none"
  )
```

## 7.4 Enrichment analysis between location
### 7.4.1 Edger
```{r table_edger, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
library(edgeR)
sample_metadata <- sample_metadata %>% arrange(sample)
counts_edger <- genome_counts_row
counts_edger=counts_edger[,which(colnames(counts_edger)%in%sample_metadata$sample)]
counts_edger <- counts_edger[,order(names(counts_edger))]

dim(sample_metadata)[1]==dim(counts_edger)[2]
colnames(counts_edger)==sample_metadata$sample
```

```{r edger, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Run EdgeR
design=model.matrix(~region,data = sample_metadata)
dge=DGEList(counts=counts_edger)
dge=estimateDisp(dge,design,robust = TRUE)
fit=glmQLFit(dge, design)
qlf=glmQLFTest(fit)

#Plot
plotMD(qlf)
abline(h=c(-1, 1), col="grey")

results=topTags(qlf, n=20)
summary(decideTests(qlf))
#13 MAGs were were found to be of higher abundance in feral than domestic cats
#1 MAG was found to be of lower abundance in feral than domestic cats
difMAGs <- results[results$table$FDR<0.05,]
taxonomy <- column_to_rownames(genome_metadata, "genome")
difMAGs <- cbind(as.data.frame(results[results$table$FDR<0.05,]), taxonomy[rownames(difMAGs),])#,table_upset_analysis_cont[rownames(difMAGs),],
#write.csv(difMAGs,"results/enrichment.csv")
```
***genome***
```{r edger_plot1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
difMAGs_row  <- difMAGs %>%
  rownames_to_column(., "genome")
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(difMAGs_row, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

difMAGs_row %>%
ggplot(aes(x=genome, y=logFC, color=phylum)) + 
  geom_point(size=4) +
  geom_hline(yintercept=0) +
  scale_color_manual(values=phylum_colors) + 
  coord_flip()+
  theme(axis.text = element_text(size = 10, face="bold.italic"),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        legend.title = element_blank(),
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))
```
***genera***
```{r edger_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
difMAGs_row %>%
ggplot(aes(x=genus, y=logFC, color=phylum)) + 
  geom_point(size=4) +
  geom_hline(yintercept=0) + 
  coord_flip()+
  scale_color_manual(values=phylum_colors)+
  theme(axis.text = element_text(size = 10, face="bold.italic"),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        legend.title = element_blank(),
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))
```

## 7.5 DESeq2
```{r deseq_lib, comment="", echo=FALSE, message=FALSE, warning=FALSE}
library(genefilter)
library(DESeq2)
```

***Genome***
```{r deseq2, comment="", echo=TRUE, message=FALSE, warning=FALSE}
diagdds = phyloseq_to_deseq2(genome_data, ~ region)

diagdds <- estimateSizeFactors(diagdds, type="poscounts",locfunc=genefilter::shorth)
diagdds = DESeq(diagdds, test="Wald", fitType="parametric")
diagdds.ins.fis <- results(diagdds, alpha=0.01, contrast=c("region", "Ittoqqortoormiit", "Daneborg"))
sigtab_diagdds.ins.fis <- diagdds.ins.fis[which(diagdds.ins.fis$pvalue < 0.05), ]
sigtab_diagdds_with_tax <- cbind(as(sigtab_diagdds.ins.fis, "data.frame"), as(tax_table(genome_data)[row.names(sigtab_diagdds.ins.fis), ], "matrix"))
#sigtab_diagdds_with_tax[order(sigtab_diagdds_with_tax$baseMean, decreasing=T), ]
deseq2_group <- as.data.frame(sigtab_diagdds_with_tax)
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
  scale_fill_brewer(palette = palname, ...)
}

x = tapply(sigtab_diagdds_with_tax$log2FoldChange, sigtab_diagdds_with_tax$order, function(x) max(x))
x = sort(x, TRUE)
sigtab_diagdds_with_tax$Family = factor(as.character(sigtab_diagdds_with_tax$order), levels=names(x))
x = tapply(sigtab_diagdds_with_tax$log2FoldChange, sigtab_diagdds_with_tax$genus, function(x) max(x))
x = sort(x, TRUE)
sigtab_diagdds_with_tax$Genus = factor(as.character(sigtab_diagdds_with_tax$genus), levels=names(x))
```

```{r deseq_plot2, comment="", echo=FALSE, message=FALSE, warning=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(sigtab_diagdds_with_tax, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

sigtab_diagdds_with_tax %>% 
  ggplot(aes(x=genome, y=log2FoldChange, color=phylum)) + 
  geom_point(size=4) +
  geom_hline(yintercept=0) + 
  coord_flip()+
  scale_color_manual(values=phylum_colors)+
  theme(axis.text = element_text(size = 8),
          axis.title = element_text(size = 12),
          legend.position = "bottom", 
          legend.title = element_blank()) +
  geom_text(aes(3, 5), label = "IT", color="#666666") +
  geom_text(aes(3, -5), label = "DB", color="#666666")
```

***Genera***
```{r deseq2_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_data_genus <- tax_glom(genome_data, 'genus')
diagdds_genus = phyloseq_to_deseq2(genome_data_genus, ~ region)

diagdds_genus <- estimateSizeFactors(diagdds_genus, type="poscounts",locfunc=genefilter::shorth)
diagdds_genus = DESeq(diagdds_genus, test="Wald", fitType="parametric")
diagdds.ins.fis_genus <- results(diagdds_genus, alpha=0.01, contrast=c("region", "Ittoqqortoormiit", "Daneborg"))
sigtab_diagdds.ins.fis_genus <- diagdds.ins.fis_genus[which(diagdds.ins.fis_genus$pvalue < 0.05), ]
sigtab_diagdds_with_tax_genus <- cbind(as(sigtab_diagdds.ins.fis_genus, "data.frame"), as(tax_table(genome_data_genus)[row.names(sigtab_diagdds.ins.fis_genus), ], "matrix"))
#sigtab_diagdds_with_tax[order(sigtab_diagdds_with_tax$baseMean, decreasing=T), ]
deseq2_group_genus <- as.data.frame(sigtab_diagdds_with_tax_genus)
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
  scale_fill_brewer(palette = palname, ...)
}

x = tapply(sigtab_diagdds_with_tax_genus$log2FoldChange, sigtab_diagdds_with_tax_genus$order, function(x) max(x))
x = sort(x, TRUE)
sigtab_diagdds_with_tax_genus$Family = factor(as.character(sigtab_diagdds_with_tax_genus$order), levels=names(x))
x = tapply(sigtab_diagdds_with_tax_genus$log2FoldChange, sigtab_diagdds_with_tax_genus$genus, function(x) max(x))
x = sort(x, TRUE)
sigtab_diagdds_with_tax_genus$Genus = factor(as.character(sigtab_diagdds_with_tax_genus$genus), levels=names(x))
```

```{r deseq_plot2_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(sigtab_diagdds_with_tax, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

sigtab_diagdds_with_tax_genus %>% 
  ggplot(aes(x=genus, y=log2FoldChange, color=phylum)) + 
  geom_point(size=4) +
  geom_hline(yintercept=0) + 
  coord_flip()+
  scale_color_manual(values=phylum_colors)+
  theme(axis.text = element_text(size = 8),
          axis.title = element_text(size = 12),
          legend.position = "bottom", 
          legend.title = element_blank()) +
  geom_text(aes(3, 5), label = "IT", color="#666666") +
  geom_text(aes(3, -5), label = "DB", color="#666666")
```
## 7.6 ANCOM-BC
```{r ancombc, comment="", echo=TRUE, message=FALSE, warning=FALSE}
phylo_counts_ancom <- genome_counts %>% 
                    filter(!genome %in% structural_zeros$genome) %>% # remove structural zeros
                    column_to_rownames("genome") %>% 
                    mutate_all(~ replace(., . == 0, 0.00001)) %>% #add pseudo counts to avoid structural zero issues 
                    otu_table(., taxa_are_rows = TRUE)

#Generate phyloseq object required to input ANCOMBC, only difference is the addition of the psudocounts
physeq_ancom <- phyloseq(phylo_counts_ancom, phylo_taxonomy, phylo_samples)
```

***Genome***
```{r ancom_rand, comment="", echo=TRUE, message=FALSE, warning=FALSE}
library(ANCOMBC)
set.seed(1234) #set seed for reproducibility
ancom_rand_output = ancombc2(data = physeq_ancom, 
                  assay_name = NULL,
                  tax_level = NULL, #change to agglomerate analysis to a higher taxonomic range
                  fix_formula = "region", #fixed variable(s)
#                  rand_formula = "(1|Individual)",
                  p_adj_method = "holm", 
                  pseudo_sens = TRUE,
                  prv_cut = 0.10, 
                  s0_perc = 0.05,
                  group = NULL, 
                  struc_zero = FALSE, 
                  neg_lb = FALSE,
                  alpha = 0.05, 
                  n_cl = 2, 
                  verbose = TRUE,
                  global = FALSE, 
                  pairwise = FALSE, 
                  dunnet = FALSE, 
                  trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  lme_control = NULL,
                  mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                  trend_control = NULL)

tax <- data.frame(physeq_ancom@tax_table) %>%
  rownames_to_column(., "taxon")

ancombc_rand_table <- ancom_rand_output$res %>%
  dplyr::select(taxon, lfc_regionIttoqqortoormiit, p_regionIttoqqortoormiit) %>%
  filter(p_regionIttoqqortoormiit < 0.05) %>%
  dplyr::arrange(p_regionIttoqqortoormiit) %>%
  merge(., tax, by="taxon")

colors_alphabetic <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", ""))  %>%
  dplyr::right_join(tax, by=join_by(phylum == phylum)) %>%
  dplyr::select(phylum, colors) %>%
  mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
	unique() %>%
	dplyr::arrange(phylum)

tax_table <- as.data.frame(unique(ancombc_rand_table$phylum))
colnames(tax_table)[1] <- "phylum"
tax_color <- merge(tax_table, colors_alphabetic, by="phylum")%>%
	dplyr::arrange(phylum) %>%
	dplyr::select(colors) %>%
	pull()
```

```{r ancombc_rand_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ggplot(ancombc_rand_table, aes(x=forcats::fct_rev(taxon), y=lfc_regionIttoqqortoormiit, color=phylum)) + 
  geom_point(size=4) + 
  scale_color_manual(values=tax_color) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  theme(panel.background = element_blank(),
          axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
          axis.title.x=element_blank())+
  xlab("Genome") + 
  ylab("log2FoldChange")+
  guides(col=guide_legend("Phylum")) +
  geom_text(aes(3, 5), label = "IT", color="#666666") +
  geom_text(aes(3, -5), label = "DB", color="#666666")
```
```{r ancom_rand_volcano, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ancom_result <- ancom_rand_output$res %>%
  na.omit() %>%
  dplyr::rename(genome=taxon) %>%
  dplyr::left_join(genome_metadata,by=join_by(genome==genome))
ancom_result %>%
    mutate(significance = ifelse(p_regionIttoqqortoormiit < 0.05, "1", "0")) %>%
    ggplot(., aes(x=-log(p_regionIttoqqortoormiit), y=lfc_regionIttoqqortoormiit, color=significance)) +
      geom_point() +
      scale_color_manual(values = c("#cccccc","#00FFFF")) +
      geom_text(aes(3, 5), label = "IT", color="#666666") +
      geom_text(aes(3, -5), label = "DB", color="#666666") +
      labs(color="Significance", y="Difference between regions", x="p-value") +
      theme_classic()
```

***Genera***
```{r ancom_rand_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
library(ANCOMBC)
set.seed(1234) #set seed for reproducibility
ancom_rand_output_genus = ancombc2(data = physeq_ancom, 
                  assay_name = NULL,
                  tax_level = "genus", #change to agglomerate analysis to a higher taxonomic range
                  fix_formula = "region", #fixed variable(s)
#                  rand_formula = "(1|Individual)",
                  p_adj_method = "holm", 
                  pseudo_sens = TRUE,
                  prv_cut = 0.10, 
                  s0_perc = 0.05,
                  group = NULL, 
                  struc_zero = FALSE, 
                  neg_lb = FALSE,
                  alpha = 0.05, 
                  n_cl = 2, 
                  verbose = TRUE,
                  global = FALSE, 
                  pairwise = FALSE, 
                  dunnet = FALSE, 
                  trend = FALSE,
                  iter_control = list(tol = 1e-5, max_iter = 20, verbose = FALSE),
                  em_control = list(tol = 1e-5, max_iter = 100),
                  lme_control = NULL,
                  mdfdr_control = list(fwer_ctrl_method = "holm", B = 100), 
                  trend_control = NULL)

tax <- data.frame(physeq_ancom@tax_table) %>%
  rownames_to_column(., "taxon") %>%
  dplyr::select(genus, phylum)

ancombc_rand_table_genus <- ancom_rand_output_genus$res %>%
  dplyr::rename(genus=taxon) %>%
  dplyr::select(genus, lfc_regionIttoqqortoormiit, p_regionIttoqqortoormiit) %>%
  filter(p_regionIttoqqortoormiit < 0.05) %>%
  dplyr::arrange(p_regionIttoqqortoormiit) %>%
  merge(., tax, by="genus")

colors_alphabetic <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", ""))  %>%
  dplyr::right_join(tax, by=join_by(phylum == phylum)) %>%
  dplyr::select(phylum, colors) %>%
  mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
	unique() %>%
	dplyr::arrange(phylum)

tax_table <- as.data.frame(unique(ancombc_rand_table_genus$phylum))
colnames(tax_table)[1] <- "phylum"
tax_color <- merge(tax_table, colors_alphabetic, by="phylum")%>%
	dplyr::arrange(phylum) %>%
	dplyr::select(colors) %>%
	pull()
```

```{r ancombc_rand_plot_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ggplot(ancombc_rand_table_genus, aes(x=forcats::fct_rev(genus), y=lfc_regionIttoqqortoormiit, color=phylum)) + 
  geom_point(size=4) + 
  scale_color_manual(values=tax_color) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  theme(panel.background = element_blank(),
          axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
          axis.title.x=element_blank())+
  xlab("Genus") + 
  ylab("log2FoldChange")+
  guides(col=guide_legend("Phylum")) +
  geom_text(aes(3, 5), label = "IT", color="#666666") +
  geom_text(aes(3, -5), label = "DB", color="#666666")
  
```
```{r ancom_rand_volcano_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ancom_result_genus <- ancom_rand_output_genus$res %>%
  na.omit() %>%
  dplyr::rename(genus=taxon) %>%
  dplyr::left_join(genome_metadata,by=join_by(genus==genus))
ancom_result_genus %>%
    mutate(significance = ifelse(p_regionIttoqqortoormiit < 0.05, "1", "0")) %>%
    ggplot(., aes(x=-log(p_regionIttoqqortoormiit), y=lfc_regionIttoqqortoormiit, color=significance)) +
      geom_point() +
      scale_color_manual(values = c("#cccccc","#00FFFF")) +
      geom_text(aes(3, 5), label = "IT", color="#666666") +
      geom_text(aes(3, -5), label = "DB", color="#666666") +
      labs(color="Significance", y="Difference between regions (LFC)", x="p-value") +
      theme_classic()
```


## 7.7 LinDA

```{r LinDA_mag, comment="", echo=TRUE, message=FALSE, warning=FALSE}
library(mia)
tse_LinDA = mia::makeTreeSummarizedExperimentFromPhyloseq(genome_data) # create tse object from the phyloseq object with structural zeros removed, no pseudo counts. 
tse_LinDA <- mia::subsetByPrevalentTaxa(tse_LinDA, detection = 0, prevalence = 0.1) # filter tse object
```

***Genome***
```{r LinDA_mag_analysis, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}

otu.tab <- as.data.frame(assay(tse_LinDA))
meta <- as.data.frame(colData(tse_LinDA)) %>% 
  dplyr::select(region)
LinDA <- LinDA::linda(
  otu.tab, 
  meta, 
  formula = '~region', 
  alpha = 0.01, 
  prev.cut = 0, # we already filtered  
  winsor.quan = 0.97)

# LinDA::linda.plot(LinDA, c('region'))
```

```{r linda_plot7, comment="", echo=FALSE, message=FALSE, warning=FALSE}
tax <- data.frame(genome_data@tax_table) %>%
  rownames_to_column(., "asv")

LinDA_table <- rownames_to_column(LinDA$output$regionIttoqqortoormiit, "asv") 
LinDA_table <- merge(LinDA_table, tax[, c("asv", "phylum", "genus")], by = "asv", all.x = TRUE)

LinDA_table_filt <- LinDA_table %>%
  dplyr::select(asv, phylum, genus, lfcSE, padj, log2FoldChange) %>%
  filter(padj < 0.05)

phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(LinDA_table_filt, by=join_by(phylum == phylum)) %>%
    arrange(match(asv,tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

ggplot(LinDA_table_filt, aes(x=forcats::fct_rev(asv), y=log2FoldChange, color=phylum)) + 
  geom_point(size=4) + 
  scale_color_manual(values=phylum_colors) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  xlab("Genome") + 
  ylab("log2FoldChange")+
  guides(col=guide_legend("Phylum")) +
  geom_text(aes(3, 5), label = "IT", color="#666666") +
  geom_text(aes(3, -5), label = "DB", color="#666666")
```
```{r LinDA_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
library(mia)
tse_LinDA_genus = mia::makeTreeSummarizedExperimentFromPhyloseq(genome_data_genus) # at genus level 
tse_LinDA_genus <- mia::subsetByPrevalentTaxa(tse_LinDA_genus, detection = 0, prevalence = 0.1) # filter tse object
```

\
***Genera***
```{r LinDA_gen_analysis, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}

otu.tab_genus <- as.data.frame(assay(tse_LinDA_genus))
meta_genus <- as.data.frame(colData(tse_LinDA_genus)) %>% 
  dplyr::select(region)
LinDA_genus <- LinDA::linda(
  otu.tab_genus, 
  meta_genus, 
  formula = '~region', 
  alpha = 0.01, 
  prev.cut = 0, # we already filtered  
  winsor.quan = 0.97)

# LinDA::linda.plot(LinDA_genus, c('region'))
```

```{r linda_plot7_gen, comment="", echo=FALSE, message=FALSE, warning=FALSE}
tax <- data.frame(genome_data_genus@tax_table) %>%
  rownames_to_column(., "asv")

LinDA_table_genus <- rownames_to_column(LinDA_genus$output$regionIttoqqortoormiit, "asv") 
LinDA_table_genus <- merge(LinDA_table_genus, tax[, c("asv", "phylum", "genus")], by = "asv", all.x = TRUE)

LinDA_table_filt_genus <- LinDA_table_genus %>%
  dplyr::select(phylum, genus, lfcSE, padj, log2FoldChange) %>%
  filter(padj < 0.05)

ggplot(LinDA_table_filt_genus, aes(x=forcats::fct_rev(genus), y=log2FoldChange, color=phylum)) + 
  geom_point(size=4) + 
  scale_color_manual(values=phylum_colors) + 
  geom_hline(yintercept=0) + 
  coord_flip()+
  xlab("Genome") + 
  ylab("log2FoldChange")+
  guides(col=guide_legend("Phylum")) +
  geom_text(aes(3, 5), label = "IT", color="#666666") +
  geom_text(aes(3, -5), label = "DB", color="#666666")
```

\
## 7.8 ALDEx2
```{r aldex, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
library(ALDEx2)
genome_counts_row <- column_to_rownames(genome_counts, "genome")
genome_counts_row_new <- genome_counts_row[,order(colnames(genome_counts_row))]
#sample_metadata_test <- sample_metadata[order(sample_metadata$sample %in% colnames(genome_counts_row),)]
sample_metadata_new <- sample_metadata[ order(sample_metadata$sample),]
colnames(genome_counts_row_new) %in% sample_metadata_new$sample
dim(sample_metadata_new)[1]==dim(genome_counts_row_new)[2]

genome_counts_aldex <- genome_counts_row_new %>% rownames_to_column(., "genome") %>%
  filter(!genome %in% structural_zeros$genome) %>% # remove structural zeros
  column_to_rownames(var="genome") %>%
  mutate_all(~ . * 1e6) %>% #multiple by a million
  round(0) #round to integer

genome_counts_filtered.clr <- aldex.clr(genome_counts_aldex, 
               sample_metadata_new$region, #not a factor
               mc.samples=128, 
               denom="all", 
               verbose=F)
```

```{r aldex_test, comment="", echo=TRUE, message=FALSE, warning=FALSE}
genome_counts_filtered.ttest <- aldex.ttest(genome_counts_filtered.clr, 
                hist.plot=F, 
                paired.test=F, 
                verbose=F)
```

```{r aldex_effect, comment="", echo=TRUE, message=FALSE, warning=FALSE}
genome_counts_filtered.effect <- aldex.effect(genome_counts_filtered.clr, 
              CI=T, 
              verbose=F, 
              include.sample.summary=F, 
              glm.conds=NULL, 
              useMC=F)
```

```{r aldex_meta, comment="", echo=TRUE, message=FALSE, warning=FALSE}
genome_counts_filtered.all <- data.frame(genome_counts_filtered.ttest,genome_counts_filtered.effect) %>%
    rownames_to_column(var="genome") %>%
    dplyr::left_join(genome_metadata,by=join_by(genome==genome))
```

```{r aldex_volcano, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_counts_filtered.all %>%
    mutate(significance = ifelse(wi.eBH < 0.05, "1", "0")) %>% 
    ggplot(., aes(x=-log(wi.eBH), y=diff.btw, color=significance)) + 
      geom_point() +
      scale_color_manual(values = c("#cccccc","#00FFFF")) +
      labs(color="Significance", y="Difference between region", x="p-value") +
      theme_classic()
```

```{r aldex_effect_plot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_counts_filtered.all %>%
    mutate(significance = ifelse(wi.eBH < 0.05, "1", "0")) %>%
    ggplot(., aes(x=diff.win, y=diff.btw, color=significance)) +
      geom_abline(intercept = 0, slope =  0, linewidth=0.4, linetype="solid",  color="#000000") +
      geom_abline(intercept = 0, slope =  1, linewidth=0.4, linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope =  2, linewidth=0.6,   linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope =  4, linewidth=0.8,   linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope = -1, linewidth=0.4, linetype="dashed", color="#000000") + 
      geom_abline(intercept = 0, slope = -2, linewidth=0.6,   linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope = -4, linewidth=0.8,   linetype="dashed", color="#000000") +
      geom_point() +
      scale_color_manual(values = c("#cccccc90","#00FFFF90")) +
      geom_text(aes(2.5, 20), label = "DB", color="#666666") +
      geom_text(aes(2.5, -20), label = "IT", color="#666666") +
      labs(color="Significance", y="Difference between region", x="Dispersion within sample types") +
      theme_classic()
```

```{r plota_padj, comment="", echo=FALSE, message=FALSE, warning=FALSE, fig.height=12, fig.width=10, fig.fullwidth=TRUE}
#Get phylum colors from the EHI standard
genome_counts_filtered.padj <- genome_counts_filtered.all %>%
    filter(wi.eBH < 0.05)

phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(genome_counts_filtered.padj, by=join_by(phylum == phylum)) %>%
    arrange(match(genome,tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

genome_counts_filtered.all %>%
    filter(wi.eBH < 0.05) %>%
    ggplot(., aes(x=effect, y=forcats::fct_rev(genome), color=phylum)) +
      geom_jitter(size=3) +
      geom_vline(xintercept=0) +
      scale_color_manual(values=phylum_colors)+ 
      xlim(-max(genome_counts_filtered.all$effect)-1,max(genome_counts_filtered.all$effect)+1)  +
      geom_text(aes(-max(genome_counts_filtered.all$effect)+1, 3.9), label = "DB", color="#666666") +
      geom_text(aes(max(genome_counts_filtered.all$effect)-1, 3.9), label = "IT", color="#666666") +
  theme(axis.text = element_text(size = 8, face="bold.italic"),
        axis.title = element_text(size = 12),
        legend.position = "right", 
        panel.background = element_blank(),
          panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "grey"))+
  xlab("log2FoldChange") + 
  ylab("Genome")
```

***Significance based on posterior probabilities***

```{r plota1, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_counts_filtered.all %>%
    mutate(significance = ifelse(overlap < 0.1, "1", "0")) %>%
    ggplot(., aes(x=diff.win, y=diff.btw, color=significance)) +
      geom_abline(intercept = 0, slope =  0, size=0.4, linetype="solid",  color="#000000") +
      geom_abline(intercept = 0, slope =  1, size=0.4, linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope =  2, size=0.6,   linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope =  4, size=0.8,   linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope = -1, size=0.4, linetype="dashed", color="#000000") + 
      geom_abline(intercept = 0, slope = -2, size=0.6,   linetype="dashed", color="#000000") +
      geom_abline(intercept = 0, slope = -4, size=0.8,   linetype="dashed", color="#000000") +
      geom_point() +
      scale_color_manual(values = c("#cccccc90","#00FFFF90")) +
      geom_text(aes(3.5, 20), label = "Enriched\nin Danenborg", color="#666666") +
      geom_text(aes(3.5, -20), label = "Enriched\nin Ittoqqortoormiit", color="#666666") +
      labs(color="Significance", y="Difference between sample types", x="Dispersion within sample types") +
      theme_classic()
```

```{r plota, comment="", echo=FALSE, message=FALSE, warning=FALSE}
genome_counts_filtered.sig <- genome_counts_filtered.all %>%
  filter(overlap < 0.1) 

#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
#  mutate_at(vars(phylum), ~ str_replace(., "[dpcofgs]__", "")) %>%
  dplyr::right_join(genome_counts_filtered.sig, by=join_by(phylum == phylum)) %>%
    arrange(match(genome,tree$tip.label)) %>%
    dplyr::select(phylum, colors) %>%
    mutate(colors = str_c(colors, "80"))  %>% #add 80% alpha
    unique() %>%
    arrange(phylum) %>%
    dplyr::select(colors) %>%
    pull()

genome_counts_filtered.sig %>%
    ggplot(., aes(x=effect, y=forcats::fct_rev(genome), color=phylum)) +
      geom_jitter(size=3) +
      geom_vline(xintercept=0) + 
      xlim(-max(genome_counts_filtered.all$effect)-1,max(genome_counts_filtered.all$effect)+1) +
      scale_color_manual(values=phylum_colors) +
      geom_text(aes(-max(genome_counts_filtered.all$effect)+1, 0.9), label = "Enriched\nin Daneborg", color="#666666") +
      geom_text(aes(max(genome_counts_filtered.all$effect)-1, 0.9), label = "Enriched\nin Ittoqqortoormiit", color="#666666") +
     theme(legend.position='right',
          panel.background = element_blank(),
          axis.line = element_line(size = 0.15, linetype = "solid", colour = "grey"),
          axis.title.x=element_blank())+
      labs(y="Genome") + 
      guides(col=guide_legend("Phylum"))
```


## 7.9 Merge Differential abundance results

```{r merge DA results, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# Wilcoxon
Wilcoxon_genome <- significant_genome %>%
  dplyr::select(., Genome, p_adjust) %>%
  dplyr::rename(
    genome = Genome,
    padj_WILCOX = p_adjust,
    )
# DESeq
DESeq_genome <- sigtab_diagdds_with_tax %>%
  dplyr::select(., genome, padj, log2FoldChange) %>%
  dplyr::rename(
    padj_DESEQ = padj,
    LogFold_DESEQ = log2FoldChange,
    )
# ANCOMBC
ANCOMBC_genome <- ancombc_rand_table %>%
  dplyr::select(., genome, p_regionIttoqqortoormiit, lfc_regionIttoqqortoormiit) %>%
  dplyr::rename(
    padj_ANCOMBC = p_regionIttoqqortoormiit,
    LogFold_ANCOMBC = lfc_regionIttoqqortoormiit,
    )
# LinDA
LinDA_genome <- LinDA_table_filt  %>%
  dplyr::select(., asv, padj, log2FoldChange)  %>%
  dplyr::rename(
    genome = asv,
    padj_LINDA = padj,
    LogFold_LINDA = log2FoldChange,
    )
#ALDEx
ALDEx_genome <- genome_counts_filtered.padj %>%
  dplyr::select(., genome, wi.eBH, effect)  %>%
  dplyr::rename(
    padj_ALDEX = wi.eBH,
    effect_ALDEX = effect,
    )
#Merge
DA_results_list <- list(Wilcoxon_genome, DESeq_genome, ANCOMBC_genome, LinDA_genome, ALDEx_genome)
DA_results<- Reduce(function(x, y) merge(x, y, all=TRUE), DA_results_list) 
DA_results <- dplyr::left_join(DA_results, genome_metadata, by = "genome")
```

\
***Upset Analysis***
```{r UpSet DA, comment="", echo=TRUE, message=FALSE, warning=FALSE}
library(UpSetR)

table_upset_DA <- DA_results  %>%
  dplyr::select(., genome, padj_WILCOX, padj_DESEQ, padj_ANCOMBC, padj_LINDA, padj_ALDEX) %>%
  dplyr::rename(
    Wilcoxon = padj_WILCOX,
    DESeq2 = padj_DESEQ, 
    ANCOMBC = padj_ANCOMBC,
    LinDA = padj_LINDA,
    ALDEx = padj_ALDEX
   ) %>%
  column_to_rownames(var="genome")  %>%
  {.[is.na(.)] <- 0; .}

table_upset_DA_binary <- ifelse(table_upset_DA > 0, 1, 0) %>% as.data.frame()
  
methodcolors=c('#c4d7d1','#408892','#c04062','#6b3a59','#e08683')

#pdf("figures/MAG_intersection.pdf",width=8,height=6, onefile=F)
upset(table_upset_DA_binary,
      sets = rev(c("Wilcoxon","DESeq2","ANCOMBC","LinDA","ALDEx")),
      sets.bar.color= rev(methodcolors),
      mb.ratio = c(0.55, 0.45), order.by = "freq")
#dev.off()
```


```{r filtered DA results, comment="", echo=TRUE, message=FALSE, warning=FALSE}
# Filtered DA Results (only MAGs found to be DA by 2+ methods)

filtered_DA_results <- DA_results %>%
  dplyr::select(., genome, padj_WILCOX, padj_DESEQ, padj_ANCOMBC, padj_LINDA, padj_ALDEX) %>%
  column_to_rownames(var="genome")  %>%
  {.[is.na(.)] <- 0; .}
filtered_DA_results <- filtered_DA_results %>%
  filter(rowSums(. > 0) >= 2) 

# Merge annotations
filtered_DA_results <- filtered_DA_results %>%
  rownames_to_column(var="genome")

filtered_DA_results <- subset(DA_results, genome %in% filtered_DA_results$genome)

filtered_DA_results <- filtered_DA_results %>%
  mutate(enrichment = ifelse(LogFold_ANCOMBC > 0 | LogFold_LINDA > 0 | LogFold_DESEQ > 0, "Ittoqqortoormiit",
                      ifelse(LogFold_LINDA < 0 | LogFold_DESEQ < 0, "Daneborg", NA)))

#library("writexl")
#write_xlsx(filtered_DA_results, "/Users/elsa/Desktop/THESIS/sled_dog_metagenomics\\DA_merged.xlsx")
```

***Mags from DA with species-level annotation***
```{r DA_species, comment="", echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
filtered_DA_results%>%
	filter(species != "s__") # 109 DA MAGs WITH species-level annotation therefore 77 DA MAGs are w/o species level annotation
```

# 8 Kraken 
```{r kraken_data, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# read in kraken data
kraken_data <- read.csv("/Users/elsa/Desktop/THESIS/sled_dog_metagenomics/resources/kraken_data.csv")

kraken_subset <- kraken_data[, c("Cetacea_.", "Pinnipedia_.",  "Eukaryota_.","Cryptosporidium_.","Giardia_.","Sarcocystis_.","Toxoplasma_gondii_.","Trichinella_.","Uncinaria_.", "Echinococcus_.", "Taenia_.", "Toxascaris_.", "Cestoda_.","Nematoda_.","Virus_.", "PARVO_Parvoviridae_.","PARVO_Bocaparvovirus_.","ADENO_Adenoviridae_.", "PAP_Papillomaviridae_.", "sample", "region")] # subset kraken_data to only include Parasites present in both regions

colNames_kraken <- setdiff(names(kraken_subset), c("region", "sample"))
```

***summary data***
```{r kraken means, comment="", echo=FALSE, message=FALSE, warning=FALSE}
column_names <-colNames_kraken
means_list_kraken <- list()

for (col_name in column_names) {
  mean_data <- kraken_data %>%
    group_by(region) %>%
    summarise(!!paste0("mean_", col_name) := mean(!!sym(col_name), na.rm = TRUE))
  
  means_list_kraken[[col_name]] <- mean_data
}

for (col_name in column_names) {
  cat("Mean for", col_name, ":\n")
  print(head(means_list_kraken[[col_name]]))
}
merged_means <- Reduce(function(x, y) merge(x, y, by = "region", all = TRUE), means_list_kraken) %>%
  knitr::kable()
```

```{r kraken_sd, comment="", echo=FALSE, message=FALSE, warning=FALSE}
sds_list_kraken <- list()

for (col_name in column_names) {
  sd_data <- kraken_data %>%
    group_by(region) %>%
    summarise(!!paste0("sd_", col_name) := sd(!!sym(col_name), na.rm = TRUE))
  
  sds_list_kraken[[col_name]] <- sd_data
}

for (col_name in column_names) {
  cat("Standard Deviation for", col_name, ":\n")
  print(head(sds_list_kraken[[col_name]]))
}

merged_sds <- Reduce(function(x, y) merge(x, y, by = "region", all = TRUE), sds_list_kraken) %>%
  knitr::kable()
```

***Wilcoxon testing (percentages)***
```{r wilcox0_kraken, comment="", echo=TRUE, message=FALSE, warning=FALSE}
wilcox_kraken <- kraken_subset %>%
    pivot_longer(-c(sample, region), names_to = "Parasite", values_to = "value") %>%
    group_by(Parasite) %>%
    summarise(p_value = {
        if (n_distinct(region) >= 2) {
            wilcox.test(value ~ region, data = cur_data(), exact = TRUE)$p.value
        } else {
            NA
        }
    }) %>%
    mutate(p_adjust = ifelse(!is.na(p_value), p.adjust(p_value, method = "BH"), NA))

significant_kraken <- wilcox_kraken %>%
  filter(!is.na(p_adjust) & p_adjust < 0.05) %>%
  knitr::kable()
```

```{r kraken plotting, comment="", echo=FALSE, message=FALSE, warning=FALSE}
for(i in colNames_kraken){
  plt_kraken <- ggplot(kraken_subset, aes(x=region, y=.data[[i]], color = region)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt_kraken)
}
```

***Fishers Exact testing (presence / absence)***
```{r inferred species, comment="", echo=FALSE, message=FALSE, warning=FALSE}
inferred_species <- read.csv("/Users/elsa/Desktop/THESIS/sled_dog_metagenomics/resources/inferred_species.csv") %>% as.data.frame()
columns_to_select <- setdiff(names(inferred_species), "region")

sample_metadata <- dplyr::left_join(sample_metadata, inferred_species %>% dplyr::select(all_of(columns_to_select)), by = "sample")
```

```{r counts of higher resolution taxonomic info, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# Parvo
count_Parvo <- aggregate(PARVO_Parvoviridae ~ region, data = sample_metadata, FUN = sum) # Daneborg = 28, Ittoq = 27
count_Bocaparvovirus <- aggregate(PARVO_Bocaparvovirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 23, Ittoq = 27
count_Human_Bocavirus <- aggregate(PARVO_Human_Bocavirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 19, Ittoq = 1
count_Bufavirus_1 <- aggregate(PARVO_Bufavirus_1 ~ region, data = sample_metadata, FUN = sum) # Daneborg 9, Ittoq = 10
count_Canine_Minute <- aggregate(PARVO_Canine_Minute ~ region, data = sample_metadata, FUN = sum) # Daneborg = 0, Ittoq = 1
count_Canine_bufavirus <- aggregate(PARVO_Canine_bufavirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 0, Ittoq = 1
count_Canine_protoparvovirus <- aggregate(PARVO_Canine_protoparvovirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 0, Ittoq = 1
count_Canine_parvovirus <- aggregate(PARVO_Canine_parvovirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 1, Ittoq = 0

# Adeno
count_Adeno <- aggregate(ADENO_Adenoviridae ~ region, data = sample_metadata, FUN = sum) # Daneborg = 19 Ittoq = 27
count_Canine_mastadenovirus <- aggregate(ADENO_Canine_mastadenovirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 0, Ittoq = 3
count_Human_adenovirus <- aggregate(ADENO_Human_adenovirus ~ region, data = sample_metadata, FUN = sum) # Daneborg = 3, Ittoq = 24

# Pap
count_Pap <- aggregate(PAP_Papillomaviridae ~ region, data = sample_metadata, FUN = sum) #Daneborg = 28 Ittoq = 29
count_HPV <- aggregate(PAP_HPV ~ region, data = sample_metadata, FUN = sum) # Daneborg = 28, Ittoq = 29
count_Tick_assoc_papillomavirus_lsx <- aggregate(PAP_Tick_assoc_papillomavirus_lsx ~ region, data = sample_metadata, FUN = sum) # Daneborg 0, Ittoq = 25
```

```{r fishers_exact, comment="", echo=FALSE, message=FALSE, warning=FALSE}
disease_columns <- c("S_capracanis", "S_tuagulusi", "S_hirsuta", "G_intestinalis", "G_peramelis", "G_muris", "Tr_spiralis", "Tr_OTHER", "Tr_pathgoniensis", "Tr_pseudospiralis", "Ta_laticollis", "Ta_asiatica", "Ta_solium", "Ta_OTHER", "Ta_krabbei", "Ta_arctos", "Ta_multiceps", "Ta_saginata", "Ta_madoquae", "E_granulosus", "E_multilocularis", "C_ubiquitum", "C_parvum", "C_hominis_TU502", "C_proventriculi", "C_muris_RN66", "C_OTHER", "C_wrairi", "C_canis", "PARVO_Parvoviridae", "PARVO_Bocaparvovirus", "PARVO_Human_Bocavirus", "PARVO_Bufavirus_1" , "PARVO_Canine_Minute", "PARVO_Canine_bufavirus", "PARVO_Canine_bufavirus", "PARVO_Canine_protoparvovirus", "PARVO_Canine_parvovirus", "ADENO_Adenoviridae", "ADENO_Human_adenovirus", "ADENO_Canine_mastadenovirus", "PAP_HPV", "PAP_Papillomaviridae", "PAP_Tick_assoc_papillomavirus_lsx")

contingency_table <- table(sample_metadata$region, sample_metadata$C_meleagridis) # change out disease to test all

# Run Fisher's exact test
fisher_result <- fisher.test(contingency_table)
print(fisher_result)
```

```{r vacc data, comment="", echo=FALSE, message=FALSE, warning=FALSE}
means_virus_vacc <- sample_metadata %>%
  group_by(Vacc_Status) %>%
  summarize(mean_virus_vacc = mean(Virus_., na.rm = TRUE)) # Not vaccinated 0.409000; Previously vaccinated 0.186875	

kraken_subset_virus <- sample_metadata[, c("Virus_.", "PARVO_Parvoviridae_.","PARVO_Bocaparvovirus_.","ADENO_Adenoviridae_.", "PAP_Papillomaviridae_.", "sample", "Vacc_Status")] # subset kraken_data to only include viruses

colNames_kraken_virus <- setdiff(names(kraken_subset_virus), c("Vacc_Status", "sample"))
```

***summary data***
```{r kraken means vacc, comment="", echo=FALSE, message=FALSE, warning=FALSE}
colNames_kraken_virus <-colNames_kraken_virus
means_list_kraken_virus <- list()

for (col_name in colNames_kraken_virus) {
  mean_data <- sample_metadata %>%
    group_by(Vacc_Status) %>%
    summarise(!!paste0("mean_", col_name) := mean(!!sym(col_name), na.rm = TRUE))
  
  means_list_kraken_virus[[col_name]] <- mean_data
}

for (col_name in colNames_kraken_virus) {
  cat("Mean for", col_name, ":\n")
  print(head(means_list_kraken_virus[[col_name]]))
}
merged_means_virus <- Reduce(function(x, y) merge(x, y, by = "Vacc_Status", all = TRUE), means_list_kraken_virus) %>%
  knitr::kable()
```

```{r kraken_sd virus, comment="", echo=FALSE, message=FALSE, warning=FALSE}
sds_list_kraken_virus <- list()

for (col_name in colNames_kraken_virus) {
  sd_data <- sample_metadata %>%
    group_by(Vacc_Status) %>%
    summarise(!!paste0("sd_", col_name) := sd(!!sym(col_name), na.rm = TRUE))
  
  sds_list_kraken_virus[[col_name]] <- sd_data
}

for (col_name in colNames_kraken_virus) {
  cat("Standard Deviation for", col_name, ":\n")
  print(head(sds_list_kraken_virus[[col_name]]))
}

merged_sds_virus <- Reduce(function(x, y) merge(x, y, by = "Vacc_Status", all = TRUE), sds_list_kraken_virus) %>%
  knitr::kable()
```

***Wilcoxon testing (percentages)***
```{r wilcox0_kraken virus, comment="", echo=TRUE, message=FALSE, warning=FALSE}
wilcox_kraken_virus <- kraken_subset_virus %>%
    pivot_longer(-c(sample, Vacc_Status), names_to = "Virus", values_to = "value") %>%
    group_by(Virus) %>%
    summarise(p_value = {
        if (n_distinct(Vacc_Status) >= 2) {
            wilcox.test(value ~ Vacc_Status, data = cur_data(), exact = TRUE)$p.value
        } else {
            NA
        }
    }) %>%
    mutate(p_adjust = ifelse(!is.na(p_value), p.adjust(p_value, method = "BH"), NA))

significant_kraken_virus <- wilcox_kraken_virus %>%
  filter(!is.na(p_adjust) & p_adjust < 0.05) %>%
  knitr::kable()

for(i in colNames_kraken_virus){
  plt_kraken_virus <- ggplot(kraken_subset_virus, aes(x=Vacc_Status, y=.data[[i]], color = Vacc_Status)) +
    geom_boxplot(alpha = 0.2, outlier.shape = NA, width = 0.3, show.legend = FALSE) +
  geom_jitter(width = 0.1, show.legend = TRUE) +
  theme_minimal() +
  theme(
    axis.line = element_line(size = 0.5, linetype = "solid", colour = "black"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank())
print(plt_kraken_virus)
}
```