---
title: "Prepare_data"
output: html_document
date: "2024-01-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE, warning=FALSE, echo=FALSE}
## Load required libraries
library(tidyverse)
library(ape)
library(distillR)
```

```{r}
read_counts <- read_tsv("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032_counts.tsv")%>%
  dplyr::rename(genome = 1)
genomes <- read_counts$genome
```

```{r}
genome_metadata <- read_tsv("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032_mag_info.tsv")%>%
  arrange(match(genome, genomes))

#%>%
#  mutate_at(vars(genome,domain,phylum,class,order,family,genus,species), ~ str_replace(., "[dpcofgs]__", "")) 
```

```{r}
genome_coverage <- read_tsv("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032_coverage.tsv")%>%
  dplyr::rename(genome = 1)%>%
  arrange(match(genome, genomes))

read_counts_row <- column_to_rownames(read_counts, "genome")
min_genome_coverage <- 0.3
read_counts_coverage <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_genome_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]]))

readlength=150 #change if sequencing read length is different
genome_counts <- read_counts_coverage %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$mag_size / readlength) ))

genome_counts <- genome_counts %>%
  column_to_rownames(., "genome") %>% 
  select(which(!colSums(., na.rm=TRUE) %in% 0))%>%
  rownames_to_column(., "genome")

genome_counts_rel <- genome_counts %>%
  mutate_at(vars(-genome), ~ . / sum(.))
```

```{r}
sample_metadata <- read_tsv("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032_metadata.tsv")%>%
	dplyr::rename(sample=EHI_plaintext) 
```


```{r}
tree <- read.tree("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032.tree")
#tree$tip.label <- str_replace_all(tree$tip.label, "'", "") # remove .fa extension
tree <- keep.tip(tree, tip = genomes) # keep only MAG tips
```

```{r func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Genome functions
kegg_table <- read_tsv("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032_merged_kegg.tsv")
```

```{r func, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Genome functions
genome_annotations <- read_tsv("https://raw.githubusercontent.com/alberdilab/sled_dog_metagenomics/main/resources/DMB0032_annotations.tsv")
```

```{r}
#Genome functions distillation
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19))
```

```{r}
save(read_counts, read_counts_coverage, genome_counts, tree, genome_metadata, kegg_table, sample_metadata, file = "/Users/jtk712/Documents/Projects/sled_dog_metagenomics/resources/data/data.Rdata") #genome_gifts,genome_annotations
```