---
title: "Data summary"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# 1. Set up working Environment

## 1.1 Load required libraries:
```{r libraries, echo=FALSE, message=FALSE, warning=FALSE}
## Load required libraries
library(ggnewscale)
library(ggtree)
library(R.utils)
library(tidyverse)
library(ape)
library(devtools)
library(ggplot2)
library(hillR)
library(spaa)
library(vegan)
library(hilldiv2) # note use of hilldiv2 rather than hilldiv
library(Rhdf5lib)
library(phyloseq)
library(phytools)
library(microbiome) 
library(matrixStats)
library(microbiomeutilities) 
library(lme4)
library(MuMIn)
library(nlme)
library(knitr) 
library(kableExtra)
library(pairwiseAdonis)
library(sjPlot)
library(distillR) 
library(RColorBrewer)
library(reshape2)
library(ggpubr)
library(ggdendro)
library(grid)
library(gplots)
library(dendextend)
library(stringr) 
library(Rtsne) 
library(glue) 
library(ggtree)
library(ggrepel)
library(ggpubr)
library(ggnewscale)
library(ggtreeExtra)
library(DESeq2)
library(ANCOMBC)
library(lefser)
library(dplyr)
library(mia)
library(pheatmap)
library(gridExtra)
```

## 1.2 Load the data
```{r loaddata, comment="", echo=FALSE, message=FALSE, warning=FALSE}
load("/Users/jtk712/Documents/Projects/sled_dog_metagenomics/resources/data/data.Rdata")
```

```{r summary, echo=FALSE, message=FALSE, warning=FALSE}
nsamples <- ncol(read_counts) # define number of samples, by reading from the counts table
metagenomic_bases <- sum(metadata$metagenomic_bases)
host_bases <- sum(metadata$host_bases)
discarded_bases <- sum(round(((metadata$metagenomic_bases+metadata$host_bases)/(1-metadata$bases_lost_fastp_percent))-(metadata$metagenomic_bases+metadata$host_bases))) #define amount of discarded bases from quality filtering
total_bases <- discarded_bases + host_bases + metagenomic_bases
singlem_bases <- sum(metadata$metagenomic_bases * metadata$singlem_fraction) 
nmags <- nrow(read_counts) # establish number of MAGs for analysis
new_species <- genome_coverage %>%
	filter(species == "s__") %>%
	nrow()

sequencing_depth <- colSums(read_counts)
sequencing_depth_sum <- sum(sequencing_depth)
sequencing_depth_mean <- mean(sequencing_depth)
sequencing_depth_sd <- sd(sequencing_depth)
```


# 2 Data pre-processing

## 2.1 General data statistics

**Number of samples**
```{r nsamples, comment="", echo=FALSE, message=FALSE, warning=FALSE}
ncol(read_counts) # 58 samples
```

**Number of MAGs**
The number of metagenome-assembled genomes (MAG) or draft bacterial genomes reconstructed from the metagenomic data.

```{r nmags, comment="", echo=FALSE, message=FALSE, warning=FALSE}
cat(nmags) # 555 MAGs
```

**Amount of total data (GB):**
The amount of total DNA data sequenced in gigabases (GB, one billion nucleotide bases).

```{r totalGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
total_bases <- discarded_bases + host_bases + metagenomic_bases
totalgb <- round(total_bases / 1000000000,2) # convert to GB 
cat(totalgb) # 333.71 GB produced
```

**Amount of discarded data (GB):**
The amount of data discarded due to low quality or lack of informativeness during data preprocesing. Discarding 5-15% of the produced data is within the expected range, due to formation of adaptor dimers, inclusion of adaptors in sequencing reads due to short insert sizes, low sequencing quality, etc.

```{r discardedGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
discardgb <- sum(round(((metadata$metagenomic_bases+metadata$host_bases)/(1-metadata$bases_lost_fastp_percent))-(metadata$metagenomic_bases+metadata$host_bases)))/1000000000
 # quality filtering
cat(discardgb) # 10.37 GB discarded
```

**Amount of discarded data (in % of the raw data):**

```{r %discarded, comment="", echo=FALSE, message=FALSE, warning=FALSE}
discarddata <- round(discarded_bases / total_bases * 100,2) 
cat(discarddata) # 3.11% removed after quality filtering
```

**Amount of host data (GB):**
The amount of data mapped against the host genome. The percentage refers to the amount of data mapped to the host genome respect to quality-filtered data. Note that this value can be very variable depending on the biological features of the sample (e.g., anal swabs contain more host DNA than faeces) and the employed reference genome (e.g., the chances for mapping to the genome are lower as the distance between) the study species and the employed reference genome differ).

```{r hostGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
hostGB <- sum(metadata$host_bases)/1000000000
cat(hostGB) # 6.49 GB from host
```

**Amount of host data (% of the quality-filtered data):**

```{r host%, comment="", echo=FALSE, message=FALSE, warning=FALSE}
hostdata <- round(host_bases / (total_bases-discarded_bases) * 100,2)
cat(hostdata) # 2.01% of quality filtered data is from host
```

**Estimated prokaryotic data:** 
The amount and proportion of data belonging to prokayotic genomes respect to the total metagenomic fraction, as estimated from singleM analysis. Note that this is an estimation that relies on the genome sizes of genomes available in reference databases. If a given taxon is not properly represented, genome size estimations can be less accurate.

```{r prokaGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
prokaGB <- sum(metadata$metagenomic_bases * metadata$singlem_fraction)/1000000000
cat(prokaGB) # 293.14 GB prokaryotic data
```

**Estimated prokaryotic data (% of the metagenomic data):** 

```{r proka%, comment="", echo=FALSE, message=FALSE, warning=FALSE}
prokadata <- round(singlem_bases / (metadata$metagenomic_bases) * 100,2)
cat(prokadata) # 92.51% of metagenomic data is prokaryotic
```

**Amount of metagenomic data (GB):**
The amount of data mapped against the host genome. The percentage refers to the amount of data mapped to the host genome respect to quality-filtered data. Note that this value can be very variable depending on the biological features of the sample (e.g., anal swabs contain more host DNA than faeces) and the employed reference genome (e.g., the chances for mapping to the genome are lower as the distance between) the study species and the employed reference genome differ).

```{r metaGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
metaGB <- sum(metadata$metagenomic_bases)/1000000000
cat(metaGB) # 316.85 GB metagenomic data
```

**Amount of metagenomic data (% of the quality-filtered data):**

```{r meta%, comment="", echo=FALSE, message=FALSE, warning=FALSE}
metaperce <- round(metagenomic_bases / (total_bases-discarded_bases) * 100,2)
cat(metaperce) # 97.99%  quality filtered data is metagenomic
```

**Total mapped sequencing depth (million reads):**
The amount of reads (and nucleotide bases) that were mapped to the entire MAG catalogue. Note that the amount of bases is only an approximation estimated by multiplying the exact number of mapped reads by 250 bp.

```{r totalreads, comment="", echo=FALSE, message=FALSE, warning=FALSE}
totalreads <- round(sequencing_depth_sum / 1000000,2)
cat(totalreads) # 1860.26 million reads
```

**Total mapped sequencing depth (GB):**

```{r mappedGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
mappedGB <- round(sequencing_depth_sum / 1000000000 * 143,2)
cat(mappedGB) # 266.02 GB total sequencing depth
```

**Average mapped sequencing depth (million reads):** 
This is the average number of reads (and nucleotide bases) mapped to each sample. Note that the amount of bases is only an approximation estimated by multiplying the exact number of mapped reads by 250 bp.
```{r meanreads, comment="", echo=FALSE, message=FALSE, warning=FALSE}
meanreads <- round(sequencing_depth_mean / 1000000,2)
cat(meanreads) # 32.07 million reads avg seq depth
```

**Average mapped sequencing depth (GB):** 
```{r meanGB, comment="", echo=FALSE, message=FALSE, warning=FALSE}
meanGB <- round(sequencing_depth_mean / 1000000000 * 143,2)
cat(meanGB) # 4.59 GB avg seq depth
```

## 2.2 General MAG statistics

**Number of MAGs without species-level annotation**
```{r}
genome_coverage %>%
    filter(species == "s__") %>%
    nrow()
```

**Number of phylums**

```{r}
genome_coverage %>%
    select(phylum) %>%
  unique() %>%
  pull() %>%
    length()
```


## 3 MAG catalogue
```{r colors, comment="", echo=FALSE, message=FALSE, warning=FALSE, results='hide'}
#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv")%>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
  arrange(match(genome, tree$tip.label)) %>%
  select(phylum, colors) %>%
  unique() %>%
  arrange(phylum) %>%
  select(colors) %>%
  pull()

# Generate the phylum color heatmap
phylum_heatmap <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv")%>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, tree$tip.label)) %>%
    select(genome,phylum) %>%
    mutate(phylum = factor(phylum, levels = unique(phylum))) %>%
    column_to_rownames(var = "genome")

# Generate  basal tree
circular_tree <- force.ultrametric(tree, method="extend") %>% # extend to ultrametric for the sake of visualisation
    ggtree(., layout = 'circular', size = 0.3, angle=45)
```
```{r circular_tree, comment="", echo=FALSE, message=FALSE, warning=FALSE}
# Add phylum ring
circular_tree <- gheatmap(circular_tree, phylum_heatmap, offset=0.85, width=0.1, colnames=FALSE) +
        scale_fill_manual(values=phylum_colors) +
        geom_tiplab2(size=1, hjust=-0.1) +
        theme(legend.position = "none", plot.margin = margin(0, 0, 0, 0), panel.margin = margin(0, 0, 0, 0))

# Flush color scale to enable a new color scheme in the next ring
circular_tree <- circular_tree + new_scale_fill()

# Add completeness ring
circular_tree <- circular_tree +
        new_scale_fill() +
        scale_fill_gradient(low = "#d1f4ba", high = "#f4baba") +
        geom_fruit(
                data=genome_metadata,
                geom=geom_bar,
                mapping = aes(x=completeness, y=genome, fill=contamination),
                offset = 0.55,
                orientation="y",
              stat="identity")

# Add genome-size ring
circular_tree <-  circular_tree +
        new_scale_fill() +
        scale_fill_manual(values = "#cccccc") +
        geom_fruit(
             data=genome_metadata,
             geom=geom_bar,
             mapping = aes(x=mag_size, y=genome),
                 offset = 0.05,
                 orientation="y",
         stat="identity")

#Plot circular tree
circular_tree
```


### 3.2 MAG details
Overview of the taxonomy and genome characteristics of the MAGs.\
**Completeness:** completeness of the MAG according to CheckM assessment.\
**Contamination:** contamination or redundancy of the MAG according to CheckM assessment.\
**Size:** size of the MAG in megabases (MB, one million nucleotide bases).
# Genome quality

```{r biplot, comment="", echo=FALSE, message=FALSE, warning=FALSE}
#Generate quality biplot
genome_biplot <- genome_metadata %>%
  select(c(genome,domain,phylum,completeness,contamination,mag_size)) %>%
  arrange(match(genome, rev(tree$tip.label))) %>% #sort MAGs according to phylogenetic tree
  ggplot(aes(x=completeness,y=contamination,size=mag_size,color=phylum)) +
              geom_point(alpha=0.7) +
                    ylim(c(10,0)) +
                    scale_color_manual(values=phylum_colors) +
                labs(y= "Contamination", x = "Completeness") +
                    theme_classic() +
                  theme(legend.position = "none")

#Generate contamination boxplot
genome_contamination <- genome_metadata %>%
            ggplot(aes(y=contamination)) +
                    ylim(c(10,0)) +
                    geom_boxplot(colour = "#999999", fill="#cccccc") +
                    theme_void() +
                    theme(legend.position = "none",
                        axis.title.x = element_blank(),
                        axis.title.y = element_blank(),
                        axis.text.y=element_blank(),
                        axis.ticks.y=element_blank(),
                        axis.text.x=element_blank(),
                        axis.ticks.x=element_blank(),
                        plot.margin = unit(c(0, 0, 0.40, 0),"inches")) #add bottom-margin (top, right, bottom, left)

#Generate completeness boxplot
genome_completeness <- genome_metadata %>%
        ggplot(aes(x=completeness)) +
                xlim(c(50,100)) +
                geom_boxplot(colour = "#999999", fill="#cccccc") +
                theme_void() +
                theme(legend.position = "none",
                    axis.title.x = element_blank(),
                    axis.title.y = element_blank(),
                    axis.text.y=element_blank(),
                    axis.ticks.y=element_blank(),
                    axis.text.x=element_blank(),
                    axis.ticks.x=element_blank(),
                    plot.margin = unit(c(0, 0, 0, 0.50),"inches")) #add left-margin (top, right, bottom, left)

#Render composite figure
grid.arrange(grobs = list(genome_completeness,genome_biplot,genome_contamination),
        layout_matrix = rbind(c(1,1,1,1,1,1,1,1,1,1,1,4),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3),
                              c(2,2,2,2,2,2,2,2,2,2,2,3)))
```

# Sequencing Assessment
## 4.1 DNA fractions

```{r}
sequence_fractions <- read_counts %>%
  pivot_longer(-genome, names_to = "sample", values_to = "value") %>%
  group_by(sample) %>%
  summarise(mags = sum(value)) %>%
    left_join(sample_metadata, by = join_by(sample == sample))  %>%
    select(sample,mags,metagenomic_bases,host_bases,bases_lost_fastp_percent) %>%
    mutate(mags_bases = mags*146) %>%
    mutate(lowqual_bases = ((metagenomic_bases+host_bases)/(1-bases_lost_fastp_percent))-(metagenomic_bases+host_bases)) %>%
    mutate(unmapped_bases = metagenomic_bases - mags_bases) %>%
    mutate(unmapped_bases = ifelse(unmapped_bases < 0, 0, unmapped_bases)) %>%
    select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases)

mags_bases_mean <- sequence_fractions %>%
    mutate(mags_bases = mags_bases / 1000000000) %>%
    select(mags_bases) %>%
    pull() %>%
    mean()
  sequence_fractions %>%
    pivot_longer(!sample, names_to = "fraction", values_to = "value") %>%
    mutate(value = value / 1000000000) %>%
    mutate(fraction = factor(fraction, levels = c("lowqual_bases","host_bases","unmapped_bases","mags_bases"))) %>%
    ggplot(., aes(x = sample, y = value, fill=fraction)) +
        geom_bar(position="stack", stat = "identity") +
            scale_fill_manual(values=c("#CCCCCC","#178a94","#ee8080","#d03161")) +
        geom_hline(yintercept = mags_bases_mean, linetype = "dashed", color = "black") +
        labs(x = "Samples", y = "Amount of data (GB)") +
        theme_classic() +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "bottom")
```

## 4.2 Estimated vs. mapped prokaryotic fraction
```{r}
singlem_table <- sequence_fractions %>%
    mutate(mags_proportion = round((mags_bases / (mags_bases + unmapped_bases))*100,2)) %>%
    left_join(sample_metadata, by = join_by(sample == sample))  %>%
    mutate(singlem_proportion = round(singlem_fraction*100,2)) %>%
    select(sample,mags_proportion,singlem_proportion) %>%
    mutate(mags_proportion = ifelse(singlem_proportion == 0, 0, mags_proportion)) %>% #convert zeros to NA
    mutate(singlem_proportion = ifelse(singlem_proportion == 0, NA, singlem_proportion)) %>% #convert zeros to NA
    mutate(singlem_proportion = ifelse(singlem_proportion < mags_proportion, NA, singlem_proportion)) %>% #if singlem is smaller, then NA, to simplify plot
    mutate(singlem_proportion = ifelse(singlem_proportion > 100, 100, singlem_proportion)) #simplify
```

```{r}
singlem_table %>%
    pivot_longer(!sample, names_to = "proportion", values_to = "value") %>%
    mutate(proportion = factor(proportion, levels = c("mags_proportion","singlem_proportion"))) %>%
    ggplot(., aes(x = value, y = sample, color=proportion)) +
            geom_line(aes(group = sample), color = "#f8a538") +
            geom_point() +
            scale_color_manual(values=c("#52e1e8","#876b53")) +
            theme_classic() +
            labs(y = "Samples", x = "Prokaryotic fraction (%)") +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "right")
```

## 4.3 Additional sequencing needed

```{r}
# Define the aimed GBs for host and mapped metagenomic data
mags_bases_aim=2
host_bases_aim=5

sequence_fractions_required <- sequence_fractions %>%
    mutate(mags_bases = round(mags_bases / 1000000000,2)) %>%
    mutate(unmapped_bases = round(unmapped_bases / 1000000000,2)) %>%
    mutate(host_bases = round(host_bases / 1000000000,2)) %>%
    mutate(lowqual_bases = round(lowqual_bases / 1000000000,2)) %>%
    mutate(total_bases = mags_bases+unmapped_bases+host_bases+lowqual_bases) %>%
    mutate(mags_bases_fraction = mags_bases/total_bases) %>%
    mutate(mags_bases_difference = mags_bases_aim - mags_bases) %>%
    mutate(meta_required = round(mags_bases_difference / mags_bases_fraction,2)) %>%
    mutate(meta_required = ifelse(meta_required < 0, 0, meta_required)) %>%
    mutate(host_bases_fraction = host_bases/total_bases) %>%
    mutate(host_bases_difference = host_bases_aim - host_bases) %>%
    mutate(host_required = round(host_bases_difference / host_bases_fraction,2)) %>%
    mutate(host_required = ifelse(host_required < 0, 0, host_required)) %>%
    select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases,meta_required,host_required)
```

```{r}
sequence_fractions_required %>%
    select(sample,meta_required,host_required) %>%
    mutate(meta_required = ifelse(meta_required > 20, 21, meta_required)) %>%
    mutate(host_required = ifelse(host_required > 20, 21, host_required)) %>%
    pivot_longer(!sample, names_to = "requirement", values_to = "value") %>%
    mutate(requirement = factor(requirement, levels = c("host_required","meta_required"))) %>%
    ggplot(., aes(x = value, y = sample, fill=requirement, group=requirement)) +
        geom_bar(position="stack", stat = "identity") +
            scale_fill_manual(values=c("#178a94","#d03161")) +
            facet_wrap(~requirement, scales="free_x") +
            labs(x = "Amount of data (GB)", y = "Samples") +
            geom_vline(xintercept = 20, linetype = "dashed", color = "black") +
        theme_classic() +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "bottom")
```










